{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'population_analysis' from '/confidential/tGenVar/vntr/output_maryam/tools/TR_analysis/scripts/population_analysis.py'>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import concurrent.futures\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import importlib\n",
    "\n",
    "import pysam\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from upsetplot import UpSet\n",
    "\n",
    "import parse_vcf_files\n",
    "importlib.reload(parse_vcf_files)\n",
    "\n",
    "import parse_fasta\n",
    "importlib.reload(parse_fasta)\n",
    "\n",
    "import exact_motif_counts\n",
    "importlib.reload(exact_motif_counts)\n",
    "\n",
    "import compare_with_asm\n",
    "importlib.reload(compare_with_asm)\n",
    "\n",
    "import plots\n",
    "importlib.reload(plots)\n",
    "\n",
    "import trio_analysis\n",
    "importlib.reload(trio_analysis)\n",
    "\n",
    "import population_analysis\n",
    "importlib.reload(population_analysis)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting plotly\n",
      "  Downloading plotly-6.0.1-py3-none-any.whl.metadata (6.7 kB)\n",
      "Collecting narwhals>=1.15.1 (from plotly)\n",
      "  Downloading narwhals-1.34.1-py3-none-any.whl.metadata (9.2 kB)\n",
      "Requirement already satisfied: packaging in /confidential/home01/Cgharegh/miniforge3/envs/TandemTwist/lib/python3.11/site-packages (from plotly) (24.1)\n",
      "Downloading plotly-6.0.1-py3-none-any.whl (14.8 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.8/14.8 MB\u001b[0m \u001b[31m85.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading narwhals-1.34.1-py3-none-any.whl (325 kB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m325.5/325.5 kB\u001b[0m \u001b[31m39.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: narwhals, plotly\n",
      "Successfully installed narwhals-1.34.1 plotly-6.0.1\n"
     ]
    }
   ],
   "source": [
    "!pip3 install seaborn\n",
    "!pip3 install upsetplot\n",
    "!pip3 install matplotlib\n",
    "!pip3 install pandas\n",
    "!pip3 install numpy\n",
    "!pip3 install pysam\n",
    "!pip3 install pyfaidx\n",
    "!pip3 install pybedtools\n",
    "!pip3 install python-Levenshtein\n",
    "!pip3 install pandarallel\n",
    "!pip3 install plotly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transform tandemTwister/vamos  format of input regions to TRGT format "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## df_TRGT_pathogenic = pd.read_csv(\"/confidential/tGenVar/Lion/TandemTwist/pathogenic_TR_TRGT.tsv\", sep=\"\\t\")\n",
    "df_TRGT_pathogenic = pd.read_csv(\"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/input_folder/regions/pathogenic_TR_TRGT.tsv\", sep=\"\\t\")\n",
    "\n",
    "df_TRGT_pathogenic.columns = ['chrom', 'start', 'end', 'ID']\n",
    "\n",
    "#chr16   66490398        66490467        ID=BEAN1;MOTIFS=TGGAA,TAAAA;STRUC=<BEAN1>\n",
    "# parse the ID field and split it into ID, MOTIFS\n",
    "df_TRGT_pathogenic['name'] = df_TRGT_pathogenic['ID'].str.split(';').str[0].str.split('=').str[1]\n",
    "# get the motifs from MOTIFS field\n",
    "df_TRGT_pathogenic['MOTIFS'] = df_TRGT_pathogenic['ID'].str.split(';').str[1].str.split('=').str[1]\n",
    "df_TRGT_pathogenic.drop('ID', axis=1, inplace=True)\n",
    "# order the columns, chrom, start, end, MOTIFS, name\n",
    "## df_TRGT_pathogenic.to_csv(\"/confidential/tGenVar/Lion/TandemTwist/pathogenic_TRs.bed\", sep=\"\\t\", index=False, header=False)\n",
    "df_TRGT_pathogenic.to_csv(\"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/input_folder/regions/pathogenic_TRs.tsv\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing the vcf file of cut asm sequence\n",
    "for HG002 (NA24385 sample in hgsvc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading the child CCS vcf file\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] Failed to open file \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi_gp1_rtr_true/TandemTwist/CCS/HG002.vcf.gz\" : No such file or directory\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] could not open variant file `b'/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi_gp1_rtr_true/TandemTwist/CCS/HG002.vcf.gz'`: No such file or directory",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 13\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# child_tandemtwister = \"/confidential/tGenVar/Lion/TandemTwist/CCS_test/HG002_tandemtwister_germline_CCS_full.vcf.gz\"\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m# father_tandemtwister = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi/TandemTwist/CCS/HG003.vcf.gz\"\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# mother_tandemtwister = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi/TandemTwist/CCS/HG004.vcf.gz\"\u001b[39;00m\n\u001b[1;32m      8\u001b[0m \n\u001b[1;32m      9\u001b[0m \u001b[38;5;66;03m# child_tandemtwister_clr = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/TandemTwist/CLR/HG002.vcf.gz\"\u001b[39;00m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m# child_tandemtwister_ont = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi/TandemTwist/ONT/HG002.vcf.gz\"\u001b[39;00m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreading the child CCS vcf file\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 13\u001b[0m child_df_tandem_twister\u001b[38;5;241m=\u001b[39m parse_vcf_files\u001b[38;5;241m.\u001b[39mparse_vcf_tandemtwister(\u001b[43mpysam\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mVariantFile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mchild_tandemtwister\u001b[49m\u001b[43m)\u001b[49m) \u001b[38;5;66;03m#, add_motif_counts=True)\u001b[39;00m\n\u001b[1;32m     14\u001b[0m \u001b[38;5;66;03m# print(\"reading the father vcf file\")\u001b[39;00m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# father_df_tandem_twister= parse_vcf_files.parse_vcf_tandemtwister(pysam.VariantFile(father_tandemtwister))\u001b[39;00m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m# print(\"reading the mother vcf file\")\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[38;5;66;03m# 510 s\u001b[39;00m\n\u001b[1;32m     26\u001b[0m \u001b[38;5;66;03m# 2.25 X faster without add_motif_counts=True\u001b[39;00m\n",
      "File \u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/pysam/libcbcf.pyx:4117\u001b[0m, in \u001b[0;36mpysam.libcbcf.VariantFile.__init__\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/pysam/libcbcf.pyx:4342\u001b[0m, in \u001b[0;36mpysam.libcbcf.VariantFile.open\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] could not open variant file `b'/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi_gp1_rtr_true/TandemTwist/CCS/HG002.vcf.gz'`: No such file or directory"
     ]
    }
   ],
   "source": [
    "### sample information\n",
    "child_sex = \"male\"\n",
    "\n",
    "child_tandemtwister = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi_gp1_rtr_true/TandemTwist/CCS/HG002.vcf.gz\"\n",
    "# child_tandemtwister = \"/confidential/tGenVar/Lion/TandemTwist/CCS_test/HG002_tandemtwister_germline_CCS_full.vcf.gz\"\n",
    "# father_tandemtwister = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi/TandemTwist/CCS/HG003.vcf.gz\"\n",
    "# mother_tandemtwister = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi/TandemTwist/CCS/HG004.vcf.gz\"\n",
    "\n",
    "# child_tandemtwister_clr = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/TandemTwist/CLR/HG002.vcf.gz\"\n",
    "# child_tandemtwister_ont = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/Ashkenazi/TandemTwist/ONT/HG002.vcf.gz\"\n",
    "\n",
    "print(\"reading the child CCS vcf file\")\n",
    "child_df_tandem_twister= parse_vcf_files.parse_vcf_tandemtwister(pysam.VariantFile(child_tandemtwister)) #, add_motif_counts=True)\n",
    "# print(\"reading the father vcf file\")\n",
    "# father_df_tandem_twister= parse_vcf_files.parse_vcf_tandemtwister(pysam.VariantFile(father_tandemtwister))\n",
    "# print(\"reading the mother vcf file\")\n",
    "# mother_df_tandem_twister= parse_vcf_files.parse_vcf_tandemtwister(pysam.VariantFile(mother_tandemtwister))\n",
    "\n",
    "# print(\"reading the child CLR vcf file\")\n",
    "# child_df_tandem_twister_clr = parse_vcf_files.parse_vcf_tandemtwister(pysam.VariantFile(child_tandemtwister_clr)) #, add_motif_counts=True)\n",
    "\n",
    "# print(\"reading the child ONT file\")\n",
    "# child_df_tandem_twister_ont = parse_vcf_files.parse_vcf_tandemtwister(pysam.VariantFile(child_tandemtwister_ont)) #, add_motif_counts=True)\n",
    "\n",
    "# 510 s\n",
    "# 2.25 X faster without add_motif_counts=True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# display the parsed vcf file, child CCS\n",
    "\n",
    "child_df_tandem_twister\n",
    "# filtered_df = child_df_tandem_twister[\n",
    "#     (child_df_tandem_twister['motif_concat_h1'].isnull()) |\n",
    "#     (child_df_tandem_twister['motif_concat_h2'].isnull())\n",
    "# ]\n",
    "# filtered_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TR statistics in HG002\n",
    "\n",
    "reproduce the previous statistics plots that we had for HG002 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_motif_lengths(row, motif_len_count):\n",
    "    # update the counts of motifs --- motif_len : count\n",
    "    if pd.isna(row['motif_ids_H1']) or pd.isna(row['motif_ids_H2']):\n",
    "        return 0\n",
    "    \n",
    "    motifs = row['motifs']\n",
    "\n",
    "    row_h1 = row['motif_ids_H1']\n",
    "    row_h2 = row['motif_ids_H2']\n",
    "    print(f'row[motif_ids_H1] = {row_h1}, type = {type(row_h1)}')\n",
    "    print(f'row[motif_ids_H1] = {row_h2}, type = {type(row_h2)}')\n",
    "\n",
    "    all_motif_ids = row['motif_ids_H1'] + row['motif_ids_H2']\n",
    "\n",
    "    for idx in all_motif_ids:\n",
    "        l = len(motifs[int(idx)])\n",
    "        if l in motif_len_count:\n",
    "            motif_len_count[l] += 1\n",
    "        else:\n",
    "            motif_len_count[l] = 1\n",
    "\n",
    "# Initialize an empty dictionary to store motif lengths\n",
    "motif_len_count = {}\n",
    "\n",
    "# Apply the function to each row of the DataFrame\n",
    "# child_df_tandem_twister.apply(get_motif_lengths, axis=1, motif_len_count=motif_len_count)\n",
    "\n",
    "# Now, motif_len_count contains the count of motif lengths\n",
    "# print(motif_len_count)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRGT output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# child_TRGT = \"/confidential/tGenVar/Lion/TandemTwist/MC/HG002_TRGT.vcf.gz\"\n",
    "child_TRGT = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/TRGT/CCS/HG002.vcf.gz\"\n",
    "# father_TRGT = \"/confidential/tGenVar/Lion/TandemTwist/MC/HG003_TRGT.vcf.gz\"\n",
    "father_TRGT = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/TRGT/CCS/HG003.vcf.gz\"\n",
    "# mother_TRGT = \"/confidential/tGenVar/Lion/TandemTwist/MC/HG004_TRGT.vcf.gz\"\n",
    "mother_TRGT = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/TRGT/CCS/HG004.vcf.gz\"\n",
    "\n",
    "\n",
    "print(\"reading child vcf\")\n",
    "child_df_tandem_TRGT= parse_vcf_files.parse_vcf_TRGT(pysam.VariantFile(child_TRGT)) #, add_motif_counts = True)\n",
    "# print(\"reading father vcf\")\n",
    "# father_df_tandem_TRGT= parse_vcf_files.parse_vcf_TRGT(pysam.VariantFile(father_TRGT))\n",
    "# print(\"reading mother vcf\")\n",
    "# mother_df_tandem_TRGT= parse_vcf_files.parse_vcf_TRGT(pysam.VariantFile(mother_TRGT))\n",
    "\n",
    "# 315 s\n",
    "# 225 s for the child only!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "child_df_tandem_TRGT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Vamos output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "child_vamos = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/vamos/CCS/HG002.vcf.gz\"\n",
    "father_vamos = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/vamos/CCS/HG003.vcf.gz\"\n",
    "mother_vamos = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/vamos/CCS/HG004.vcf.gz\"\n",
    "\n",
    "\n",
    "child_vamos_clr = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/vamos/CLR/HG002.vcf.gz\"\n",
    "child_vamos_ont = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/Ashkenazi/vamos/ONT/HG002.vcf.gz\"\n",
    "\n",
    "# Funktion zum Komprimieren und Indexieren einer Datei\n",
    "def compress_and_index(file_path):\n",
    "    # Komprimieren mit bgzip\n",
    "    subprocess.run(['bgzip', '-c', file_path], stdout=open(file_path + '.gz', 'wb'))\n",
    "    # Indexieren mit tabix\n",
    "    subprocess.run(['tabix', '-p', 'vcf', file_path + '.gz'])\n",
    "\n",
    "\n",
    "\n",
    "print(\"reading child vcf\")\n",
    "child_df_vamos = parse_vcf_files.parse_vcf_vamos(pysam.VariantFile(child_vamos)) #, add_motif_counts = True)\n",
    "# print(\"reading father vcf\")\n",
    "# father_df_vamos = parse_vcf_files.parse_vcf_vamos(pysam.VariantFile(father_vamos))\n",
    "# print(\"reading mother vcf\")\n",
    "# mother_df_vamos = parse_vcf_files.parse_vcf_vamos(pysam.VariantFile(mother_vamos))\n",
    "\n",
    "# print(\"reading child vcf clr\")\n",
    "# child_df_vamos_clr = parse_vcf_files.parse_vcf_vamos(pysam.VariantFile(child_vamos_clr)) #, add_motif_counts = True)\n",
    "\n",
    "# print(\"reading child vcf ont\")\n",
    "# child_df_vamos_ont = parse_vcf_files.parse_vcf_vamos(pysam.VariantFile(child_vamos_clr)) #, add_motif_counts = True)\n",
    "\n",
    "# 851 s\n",
    "# 140 s for one sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "child_df_vamos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comapre with assembly sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first read the assembly sample. Store it to a dataframe with these columns:\n",
    "chrom, start, end, asm_h1_seq, asm_h2_seq...\n",
    "It has the cut sequence of the assembly sample in each region in the two haplotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>motifs</th>\n",
       "      <th>motif_ids</th>\n",
       "      <th>CN</th>\n",
       "      <th>CN_ref</th>\n",
       "      <th>GT</th>\n",
       "      <th>len_</th>\n",
       "      <th>motif_concat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>94809</td>\n",
       "      <td>94847</td>\n",
       "      <td>(T,)</td>\n",
       "      <td>(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>95539</td>\n",
       "      <td>95569</td>\n",
       "      <td>(AAC, AAT, ATC, AGC)</td>\n",
       "      <td>(0, 0, 1, 0, 1, 0, 0, 0, 3, 0)</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>AACAACAATAACAATAACAACAACAGCAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>381155</td>\n",
       "      <td>381699</td>\n",
       "      <td>(GAGATGCCCAGGCCTGGCAGCCGGCGCACGTGGGTTCTCTGTGGC...</td>\n",
       "      <td>(4, 16, 4, 18, 4, 18, 4, 19)</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>545</td>\n",
       "      <td>GGAGATGCCCAGGCCTGGCGGCCGGCGCACGCGGGTTCTCTGTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>382245</td>\n",
       "      <td>382665</td>\n",
       "      <td>(GTGAGAACGTGTGTGTAGTGTCCACATGTCCTCTGTGC, GTGAG...</td>\n",
       "      <td>(10, 6, 15, 12, 0, 15, 18, 6, 15, 1, 10, 6)</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>419</td>\n",
       "      <td>GTGAGTTCATGGGTGTGACGGGGTGTGTGCTGTGTGAGAACGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>382990</td>\n",
       "      <td>383050</td>\n",
       "      <td>(AAAAATTTTT, AAATCTTT, AAAAATT, TTAAATT, TTTCTTT)</td>\n",
       "      <td>(2, 0, 1, 2, 2, 3, 4, 0)</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>AAAAATTAAAAATTTTTAAATCTTTAAAAATTAAAAATTTTAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1211825</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155996508</td>\n",
       "      <td>155997070</td>\n",
       "      <td>(GGTGGCTCGCACCTGTAATCCCAGCACTTAGGGAGGCTTAGGTGG...</td>\n",
       "      <td>(7, 14)</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>561</td>\n",
       "      <td>GGTGGCTCGCACCTGTAATCCCAGCACTTAGGGAGGCTTAGGTGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1211826</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156000142</td>\n",
       "      <td>156000183</td>\n",
       "      <td>(AT, AC, AA)</td>\n",
       "      <td>(2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, ...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>AAATATATATATATATATATACACATATATATATATAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1211827</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156001196</td>\n",
       "      <td>156001225</td>\n",
       "      <td>(CCTGTG, CCTGG, GTTAG, GCTGG)</td>\n",
       "      <td>(1, 2, 0, 3, 1)</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>CCTGGGTTAGCCTGTGGCTGGCCTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1211828</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156002754</td>\n",
       "      <td>156002803</td>\n",
       "      <td>(TTAGGACACAGGACACTGTGTGAGTG, TTAGGACACAGGACACT...</td>\n",
       "      <td>(2, 1)</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>49</td>\n",
       "      <td>AGGACGCTGGACACTGTGTGAGTGTTAGGACACAGGACACTGTGTGAGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1211829</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156002944</td>\n",
       "      <td>156002981</td>\n",
       "      <td>(GGCTCCT, GGCACCT, GCATCT, GCACCT, GGCCT)</td>\n",
       "      <td>(1, 0, 4, 2, 2, 3)</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>GGCACCTGGCTCCTGGCCTGCATCTGCATCTGCACCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1211830 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end  \\\n",
       "0        chr1      94809      94847   \n",
       "1        chr1      95539      95569   \n",
       "2        chr1     381155     381699   \n",
       "3        chr1     382245     382665   \n",
       "4        chr1     382990     383050   \n",
       "...       ...        ...        ...   \n",
       "1211825  chrX  155996508  155997070   \n",
       "1211826  chrX  156000142  156000183   \n",
       "1211827  chrX  156001196  156001225   \n",
       "1211828  chrX  156002754  156002803   \n",
       "1211829  chrX  156002944  156002981   \n",
       "\n",
       "                                                    motifs  \\\n",
       "0                                                     (T,)   \n",
       "1                                     (AAC, AAT, ATC, AGC)   \n",
       "2        (GAGATGCCCAGGCCTGGCAGCCGGCGCACGTGGGTTCTCTGTGGC...   \n",
       "3        (GTGAGAACGTGTGTGTAGTGTCCACATGTCCTCTGTGC, GTGAG...   \n",
       "4        (AAAAATTTTT, AAATCTTT, AAAAATT, TTAAATT, TTTCTTT)   \n",
       "...                                                    ...   \n",
       "1211825  (GGTGGCTCGCACCTGTAATCCCAGCACTTAGGGAGGCTTAGGTGG...   \n",
       "1211826                                       (AT, AC, AA)   \n",
       "1211827                      (CCTGTG, CCTGG, GTTAG, GCTGG)   \n",
       "1211828  (TTAGGACACAGGACACTGTGTGAGTG, TTAGGACACAGGACACT...   \n",
       "1211829          (GGCTCCT, GGCACCT, GCATCT, GCACCT, GGCCT)   \n",
       "\n",
       "                                                 motif_ids  CN  CN_ref GT  \\\n",
       "0        (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  35      35  0   \n",
       "1                           (0, 0, 1, 0, 1, 0, 0, 0, 3, 0)  10      10  1   \n",
       "2                             (4, 16, 4, 18, 4, 18, 4, 19)   8       8  1   \n",
       "3              (10, 6, 15, 12, 0, 15, 18, 6, 15, 1, 10, 6)  12      12  1   \n",
       "4                                 (2, 0, 1, 2, 2, 3, 4, 0)   8       8  1   \n",
       "...                                                    ...  ..     ... ..   \n",
       "1211825                                            (7, 14)   2       2  1   \n",
       "1211826  (2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, ...  19      19  1   \n",
       "1211827                                    (1, 2, 0, 3, 1)   5       5  0   \n",
       "1211828                                             (2, 1)   2       2  1   \n",
       "1211829                                 (1, 0, 4, 2, 2, 3)   6       6  0   \n",
       "\n",
       "         len_                                       motif_concat  \n",
       "0          35                TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT  \n",
       "1          30                     AACAACAATAACAATAACAACAACAGCAAC  \n",
       "2         545  GGAGATGCCCAGGCCTGGCGGCCGGCGCACGCGGGTTCTCTGTGGC...  \n",
       "3         419  GTGAGTTCATGGGTGTGACGGGGTGTGTGCTGTGTGAGAACGTGTG...  \n",
       "4          63  AAAAATTAAAAATTTTTAAATCTTTAAAAATTAAAAATTTTAAATT...  \n",
       "...       ...                                                ...  \n",
       "1211825   561  GGTGGCTCGCACCTGTAATCCCAGCACTTAGGGAGGCTTAGGTGGG...  \n",
       "1211826    38             AAATATATATATATATATATACACATATATATATATAT  \n",
       "1211827    26                         CCTGGGTTAGCCTGTGGCTGGCCTGG  \n",
       "1211828    49  AGGACGCTGGACACTGTGTGAGTGTTAGGACACAGGACACTGTGTGAGT  \n",
       "1211829    37              GGCACCTGGCTCCTGGCCTGCATCTGCATCTGCACCT  \n",
       "\n",
       "[1211830 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from parse_vcf_files import parse_asm_vcf_tandemtwister\n",
    "\n",
    "NA19240_h1 = \"/Users/maryamghareghani/Documents/molgen_mount/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/NA19240_h1.vcf.gz\"\n",
    "\n",
    "NA19240_h1_df = parse_asm_vcf_tandemtwister(pysam.VariantFile(NA19240_h1))\n",
    "\n",
    "NA19240_h1_df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info = pd.read_csv('/Users/maryamghareghani/Documents/molgen_mount/tools/run_all_tools/input_folder/samples/hgsvc/HGSVC2_sample_info.tsv', sep='\\t')\n",
    "h2 = sample_info.copy()\n",
    "h2['sample'] = h2['sample'] + '_h2'\n",
    "\n",
    "sample_info['sample'] = sample_info['sample'] + '_h1'\n",
    "\n",
    "# NA19249_h1_df\n",
    "\n",
    "sample_info = pd.concat([sample_info, h2], ignore_index=True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00268_h1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00512_h1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00733_h1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00732_h1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00864_h1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>NA19983_h2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>HG03683_h2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>NA19650_h2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>NA19434_h2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>NA20847_h2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample\n",
       "0   HG00268_h1\n",
       "1   HG00512_h1\n",
       "2   HG00733_h1\n",
       "3   HG00732_h1\n",
       "4   HG00864_h1\n",
       "..         ...\n",
       "83  NA19983_h2\n",
       "84  HG03683_h2\n",
       "85  NA19650_h2\n",
       "86  NA19434_h2\n",
       "87  NA20847_h2\n",
       "\n",
       "[88 rows x 1 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NA19249_h1_df.merge(sample_info, on='sample', how='inner')\n",
    "\n",
    "sample_info[sample_info['sample']=='NA19249_h1']\n",
    "\n",
    "sample_info[['sample']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# African trio\n",
    "# child   mother  father  super.pop\n",
    "# HG00733 HG00732 HG00731 AMR\n",
    "# HG00514 HG00513 HG00512 EAS\n",
    "# NA19240 NA19238 NA19239 AFR\n",
    "\n",
    "import trio_analysis\n",
    "importlib.reload(trio_analysis)\n",
    "\n",
    "from parse_vcf_files import parse_asm_vcf_tandemtwister\n",
    "\n",
    "trio_dict_afr = {\n",
    "    \"child_h1_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/NA19240_h1.vcf.gz\",\n",
    "    \"child_h2_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/NA19240_h2.vcf.gz\",\n",
    "    \"mother_h1_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/NA19238_h1.vcf.gz\",\n",
    "    \"mother_h2_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/NA19238_h2.vcf.gz\",\n",
    "    \"father_h1_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/NA19239_h1.vcf.gz\",\n",
    "    \"father_h2_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/NA19239_h2.vcf.gz\"\n",
    "}\n",
    "\n",
    "trio_dict_amr = {\n",
    "    \"child_h1_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00733_h1.vcf.gz\",\n",
    "    \"child_h2_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00733_h2.vcf.gz\",\n",
    "    \"mother_h1_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00732_h1.vcf.gz\",\n",
    "    \"mother_h2_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00732_h2.vcf.gz\",\n",
    "    \"father_h1_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00731_h1.vcf.gz\",\n",
    "    \"father_h2_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00731_h2.vcf.gz\"\n",
    "}\n",
    "\n",
    "trio_dict_eas = {\n",
    "    \"child_h1_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00514_h1.vcf.gz\",\n",
    "    \"child_h2_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00514_h2.vcf.gz\",\n",
    "    \"mother_h1_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00513_h1.vcf.gz\",\n",
    "    \"mother_h2_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00513_h2.vcf.gz\",\n",
    "    \"father_h1_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00512_h1.vcf.gz\",\n",
    "    \"father_h2_file\" : \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output_not_complete/hgsvc/TandemTwist/asm/HG00512_h2.vcf.gz\"\n",
    "}\n",
    "\n",
    "print('analyzing AMR trio')\n",
    "trio_df_amr = trio_analysis.compute_trio_inheritance_df(trio_dict_amr)\n",
    "trio_df_amr.to_csv('output/trio_df_amr.csv', index = False)\n",
    "\n",
    "print('Analyzing EAS trio')\n",
    "trio_df_eas = trio_analysis.compute_trio_inheritance_df(trio_dict_eas)\n",
    "trio_df_eas.to_csv('output/trio_df_eas.csv', index = False)\n",
    "\n",
    "\n",
    "# trio_df_afr = trio_analysis.compute_trio_inheritance_df(trio_dict)\n",
    "# trio_df_afr.to_csv('output/trio_df_afr.csv', index = False)\n",
    "# trio_df_afr = pd.read_csv('output/trio_df_afr.csv')\n",
    "\n",
    "\n",
    "\n",
    "import plots\n",
    "importlib.reload(plots)\n",
    "\n",
    "plt_afr = plots.plot_karyotype_trio_hap_blocks(trio_df_afr, output=\"output/karyotype_plot_trio_afr_dpi300.pdf\", title=\"Haplotype blocks for African trio\")\n",
    "\n",
    "plt_afr.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# trio_df.drop(columns = ['hap_value_h1', 'hap_value_h2'])\n",
    "\n",
    "trio_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Get the frequency of each unique value in 'inheritance_h1'\n",
    "frequency_h1 = trio_df[(trio_df['chrom']!='chr10') & (trio_df['chrom']!='chrX')]['inheritance_h1'].value_counts()\n",
    "\n",
    "# Get the frequency of each unique value in 'inheritance_h2'\n",
    "frequency_h2 = trio_df[(trio_df['chrom']!='chr10') & (trio_df['chrom']!='chrX')]['inheritance_h2'].value_counts()\n",
    "\n",
    "print(\"Frequency of unique values in 'inheritance_h1':\")\n",
    "print(frequency_h1)\n",
    "\n",
    "print(\"\\nFrequency of unique values in 'inheritance_h2':\")\n",
    "print(frequency_h2)\n",
    "\n",
    "trio_df[(trio_df['chrom']!='chr10') & (trio_df['chrom']!='chrX') & (trio_df['inheritance_h1']=='father_denovo')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Population analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samples = ['NA18939_h2' 'NA18534_h2' 'HG00171_h1' 'NA20847_h2' 'NA20509_h2'\n",
      " 'HG00732_h1' 'NA19238_h2' 'NA18534_h1' 'NA12329_h2' 'NA19983_h1'\n",
      " 'NA19239_h2' 'HG00096_h2' 'HG03065_h1' 'HG03371_h2' 'HG02492_h2'\n",
      " 'HG03371_h1' 'NA20509_h1' 'NA19239_h1' 'HG00731_h1' 'NA20847_h1'\n",
      " 'NA19650_h2' 'HG00171_h2' 'HG03732_h2' 'NA18939_h1' 'HG02011_h2'\n",
      " 'NA19238_h1' 'HG00512_h2' 'HG01505_h2' 'NA19240_h1' 'HG03683_h2'\n",
      " 'NA19240_h2' 'NA12329_h1' 'HG02492_h1' 'NA19983_h2' 'HG00732_h2'\n",
      " 'HG02011_h1' 'HG03065_h2' 'HG01505_h1' 'HG00512_h1' 'HG00513_h1'\n",
      " 'HG00513_h2' 'HG00096_h1' 'HG03683_h1' 'HG03732_h1' 'NA19650_h1'\n",
      " 'HG00864_h2' 'HG00514_h1' 'HG00731_h2' 'HG03009_h1' 'HG01596_h2'\n",
      " 'HG03486_h1' 'HG03486_h2' 'NA12878_h1' 'HG00733_h2' 'HG01114_h1'\n",
      " 'HG03125_h2' 'HG02818_h1' 'HG02818_h2' 'HG02587_h2' 'HG00864_h1'\n",
      " 'HG01114_h2' 'HG00514_h2' 'HG03009_h2' 'HG01596_h1' 'HG02587_h1'\n",
      " 'HG03125_h1' 'NA12878_h2' 'HG00733_h1'], num = 68\n"
     ]
    }
   ],
   "source": [
    "# df[df[\"sample\"]==\"NA24385_h2\"]\n",
    "# sample_info[sample_info[\"sample\"]==\"NA24385_h2\"]\n",
    "# sample_info\n",
    "\n",
    "# with pd.option_context('display.max_rows', 88):\n",
    "#     display(sample_info)\n",
    "\n",
    "\n",
    "samples = df['sample'].unique()\n",
    "print(f'samples = {samples}, num = {len(samples)}')\n",
    "\n",
    "# present samples\n",
    "# samples = ['NA18939_h2' 'NA18534_h2' 'HG00171_h1' 'NA20847_h2' 'NA20509_h2'\n",
    "#  'HG00732_h1' 'NA19238_h2' 'NA18534_h1' 'NA12329_h2' 'NA19983_h1'\n",
    "#  'NA19239_h2' 'HG00096_h2' 'HG03065_h1' 'HG03371_h2' 'HG02492_h2'\n",
    "#  'HG03371_h1' 'NA20509_h1' 'NA19239_h1' 'HG00731_h1' 'NA20847_h1'\n",
    "#  'NA19650_h2' 'HG00171_h2' 'HG03732_h2' 'NA18939_h1' 'HG02011_h2'\n",
    "#  'NA19238_h1' 'HG00512_h2' 'HG01505_h2' 'NA19240_h1' 'HG03683_h2'\n",
    "#  'NA19240_h2' 'NA12329_h1' 'HG02492_h1' 'NA19983_h2' 'HG00732_h2'\n",
    "#  'HG02011_h1' 'HG03065_h2' 'HG01505_h1' 'HG00512_h1' 'HG00513_h1'\n",
    "#  'HG00513_h2' 'HG00096_h1' 'HG03683_h1' 'HG03732_h1' 'NA19650_h1'\n",
    "#  'HG00864_h2' 'HG00514_h1' 'HG00731_h2' 'HG03009_h1' 'HG01596_h2'\n",
    "#  'HG03486_h1' 'HG03486_h2' 'NA12878_h1' 'HG00733_h2' 'HG01114_h1'\n",
    "#  'HG03125_h2' 'HG02818_h1' 'HG02818_h2' 'HG02587_h2' 'HG00864_h1'\n",
    "#  'HG01114_h2' 'HG00514_h2' 'HG03009_h2' 'HG01596_h1' 'HG02587_h1'\n",
    "#  'HG03125_h1' 'NA12878_h2' 'HG00733_h1'], num = 68"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'cn_allele_count_summary' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[54], line 8\u001b[0m\n\u001b[1;32m      2\u001b[0m importlib\u001b[38;5;241m.\u001b[39mreload(population_analysis)\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# plt_alleles_cn = population_analysis.plot_allele_count(cn_allele_count_summary)\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# plt_alleles_len = population_analysis.plot_allele_count(len_allele_count_summary)\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m# plt_alleles_motif = population_analysis.plot_allele_count(motif_allele_count_summary)\u001b[39;00m\n\u001b[0;32m----> 8\u001b[0m plt \u001b[38;5;241m=\u001b[39m population_analysis\u001b[38;5;241m.\u001b[39mplot_allele_count(\u001b[43mcn_allele_count_summary\u001b[49m, len_allele_count_summary, motif_allele_count_summary)\n\u001b[1;32m     10\u001b[0m plt\u001b[38;5;241m.\u001b[39msavefig(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(path, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpopulation\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnum_alleles_barplots.pdf\u001b[39m\u001b[38;5;124m'\u001b[39m), \u001b[38;5;28mformat\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpdf\u001b[39m\u001b[38;5;124m'\u001b[39m, bbox_inches\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtight\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'cn_allele_count_summary' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "import population_analysis\n",
    "importlib.reload(population_analysis)\n",
    "\n",
    "# plt_alleles_cn = population_analysis.plot_allele_count(cn_allele_count_summary)\n",
    "# plt_alleles_len = population_analysis.plot_allele_count(len_allele_count_summary)\n",
    "# plt_alleles_motif = population_analysis.plot_allele_count(motif_allele_count_summary)\n",
    "\n",
    "plt = population_analysis.plot_allele_count(cn_allele_count_summary, len_allele_count_summary, motif_allele_count_summary)\n",
    "\n",
    "plt.savefig(os.path.join(path, 'population', 'num_alleles_barplots.pdf'), format='pdf', bbox_inches='tight')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Population analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading sample info\n",
      "all samples: ['HG00733_h1', 'HG00733_h2', 'HG00731_h1', 'HG00731_h2', 'HG00732_h1', 'HG00732_h2']\n",
      "processing sample 0: HG00731_h1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/local/ipykernel_85255/1753075323.py:35: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sample_superpop['superpop'] = pd.Categorical(sample_superpop['superpop'], categories=superpop_order, ordered=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing sample 1: HG00731_h2\n",
      "processing sample 2: HG00732_h1\n",
      "processing sample 3: HG00732_h2\n",
      "processing sample 4: HG00733_h1\n",
      "processing sample 5: HG00733_h2\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import importlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import population_analysis_efficient\n",
    "importlib.reload(population_analysis_efficient)\n",
    "import time\n",
    "\n",
    "time_ = time.time()\n",
    "\n",
    "directory = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/hgsvc_small_test/TandemTwist/asm/\"\n",
    "sample_info_file = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/input_folder/samples/hgsvc/HGSVC2_sample_info.tsv\"\n",
    "output_file = 'output/pop_efficient/population_df_small_test.csv'\n",
    "superpop_order=['AFR', 'AMR', 'EUR', 'EAS', 'SAS']\n",
    "sample_order = None\n",
    "col_prefix = 'seq'\n",
    "\n",
    "#df_pop = population_analysis_efficient.compute_population_df(directory=directory, output=output_file)\n",
    "# 96m for all samples\n",
    "\n",
    "df = pd.read_csv(output_file, sep = '\\t')\n",
    "\n",
    "# commulative allele classification\n",
    "\n",
    "sample_info = population_analysis_efficient.get_sample_info_df(sample_info_file)\n",
    "\n",
    "all_samples = [col.replace('CN_', '') for col in df.columns if col.startswith('CN_') and col != 'CN_ref']\n",
    "print('all samples:', all_samples)\n",
    "\n",
    "subsample_info = sample_info[sample_info['sample'].isin(all_samples)]\n",
    "\n",
    "sample_superpop = subsample_info[['sample', 'superpop']]\n",
    "\n",
    "if superpop_order:\n",
    "    sample_superpop['superpop'] = pd.Categorical(sample_superpop['superpop'], categories=superpop_order, ordered=True)\n",
    "sample_superpop = sample_superpop.sort_values(by=['superpop', 'sample'])\n",
    "if sample_order is None:\n",
    "    sample_order = sample_superpop['sample'].tolist()\n",
    "\n",
    "\n",
    "# create two dataframes for computing allele sets and counting the number of singleton, biallelic, and multiallelic\n",
    "df_alleles_set = df[[\"chrom\", \"start\", \"end\"]].copy()\n",
    "df_num_alleles = df[[\"chrom\", \"start\", \"end\"]].copy()\n",
    "\n",
    "# computing and counting the set of alleles seen up to each added sample\n",
    "prev_sample = None\n",
    "for i, sample in enumerate(sample_order):\n",
    "\n",
    "    print(f'processing sample {i}: {sample}')\n",
    "\n",
    "    if df_alleles_set.shape[1] == 3:\n",
    "        # no sample is added yet\n",
    "        df_alleles_set.loc[:, f\"{sample}\"] = df[f'{col_prefix}_{sample}'].apply(lambda x: set([x]) if pd.notna(x) else set())\n",
    "    \n",
    "    else:\n",
    "        df_alleles_set.loc[:, f\"{sample}\"] = df_alleles_set[f\"{prev_sample}\"].combine(df[f'{col_prefix}_{sample}'], \n",
    "                                                                        lambda s1, s2: s1.union(set([s2])) if pd.notna(s2) else s1)\n",
    "    \n",
    "    df_num_alleles.loc[:, f\"{sample}\"] = df_alleles_set[f\"{sample}\"].apply(len)\n",
    "\n",
    "    prev_sample = sample\n",
    "\n",
    "\n",
    "\n",
    "# 3m 30s for 6 samples (for all CN, len, and seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PCA analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import importlib\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import LabelEncoder, StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "import plotly.express as px\n",
    "\n",
    "import population_analysis_efficient\n",
    "importlib.reload(population_analysis_efficient)\n",
    "from population_analysis_efficient import compute_pca\n",
    "\n",
    "pop_df_file = 'output/pop_efficient/population_df.csv'\n",
    "sample_info_file = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/input_folder/samples/hgsvc/HGSVC2_sample_info.tsv\"\n",
    "\n",
    "std_threshold = 0  # filter regions with std <= threshold\n",
    "\n",
    "# sample_info = population_analysis_efficient.get_sample_info_df(sample_info_file)\n",
    "# df = pd.read_csv(pop_df_file, sep = '\\t')\n",
    "\n",
    "pca_df = compute_pca(df, sample_info, value_col='seq', std_threshold=std_threshold)\n",
    "# 13 s, 19 s\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>motifs</th>\n",
       "      <th>CN_ref</th>\n",
       "      <th>CN_NA18939_h2</th>\n",
       "      <th>len_NA18939_h2</th>\n",
       "      <th>seq_NA18939_h2</th>\n",
       "      <th>CN_NA18534_h2</th>\n",
       "      <th>len_NA18534_h2</th>\n",
       "      <th>...</th>\n",
       "      <th>seq_HG03125_h1</th>\n",
       "      <th>CN_NA12878_h2</th>\n",
       "      <th>len_NA12878_h2</th>\n",
       "      <th>seq_NA12878_h2</th>\n",
       "      <th>CN_HG00733_h1</th>\n",
       "      <th>len_HG00733_h1</th>\n",
       "      <th>seq_HG00733_h1</th>\n",
       "      <th>CN_NA24385_h2</th>\n",
       "      <th>len_NA24385_h2</th>\n",
       "      <th>seq_NA24385_h2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>819914</td>\n",
       "      <td>823663</td>\n",
       "      <td>('ACTACCCTATACAGCAGA', 'AGGTCCACATTGTCCAGC', '...</td>\n",
       "      <td>212</td>\n",
       "      <td>223</td>\n",
       "      <td>3910</td>\n",
       "      <td>CTGTCGAGCAGATCCACCCTGTCTATACTACCTGACTCTCCAGAAG...</td>\n",
       "      <td>223</td>\n",
       "      <td>3910</td>\n",
       "      <td>...</td>\n",
       "      <td>CTGTCGAGCAGATCCACCCTGTCTATACTACCTGACTCTCCAGAAG...</td>\n",
       "      <td>223</td>\n",
       "      <td>3910</td>\n",
       "      <td>CTGTCGAGCAGATCCACCCTGTCTATACTACCTGACTCTCCAGAAG...</td>\n",
       "      <td>223</td>\n",
       "      <td>3910</td>\n",
       "      <td>CTGTCGAGCAGATCCACCCTGTCTATACTACCTGACTCTCCAGAAG...</td>\n",
       "      <td>224</td>\n",
       "      <td>3919</td>\n",
       "      <td>CTGTCGAGCAGATCCACCCTGTCTATACTACCTGCCTCTCCAGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>827584</td>\n",
       "      <td>827629</td>\n",
       "      <td>('GCCGGGGCGGGGT', 'GCGGGGCCGGGCCGG', 'GGCGGGCT...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>43</td>\n",
       "      <td>GCGGGGCCGGGCCGGGCGGGGCCGGGCCGGGCCGGGGCGGGGT</td>\n",
       "      <td>3</td>\n",
       "      <td>43</td>\n",
       "      <td>...</td>\n",
       "      <td>GCGGGGCCGGGCCGGGCGGGGCCGGGCCGGGCCGGGGCGGGGT</td>\n",
       "      <td>3</td>\n",
       "      <td>43</td>\n",
       "      <td>GCGGGGCCGGGCCGGGCGGGGCCGGGCCGGGCCGGGGCGGGGT</td>\n",
       "      <td>3</td>\n",
       "      <td>43</td>\n",
       "      <td>GCGGGGCCGGGCCGGGCGGGGCCGGGCCGGGCCGGGGCGGGGT</td>\n",
       "      <td>3</td>\n",
       "      <td>43</td>\n",
       "      <td>GCGGGGCCGGGCCGGGCCGGGGCGGGGTGCGGGGCCGGGCCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>829414</td>\n",
       "      <td>829478</td>\n",
       "      <td>('ATGCA', 'ATGCAT', 'GGGT', 'GGT', 'GTACT', 'G...</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>57</td>\n",
       "      <td>GTGTGTGTGTGTGTGCTGTACTGTGCTGTGCTATGCATGTATTATG...</td>\n",
       "      <td>12</td>\n",
       "      <td>57</td>\n",
       "      <td>...</td>\n",
       "      <td>GTGTGTGTGTGTGTGCTGTACTGTGCTGTGCTATGCATGTATTATG...</td>\n",
       "      <td>12</td>\n",
       "      <td>57</td>\n",
       "      <td>GTGTGTGTGTGTGTGCTGTACTGTGCTGTGCTATGCATGTATTATG...</td>\n",
       "      <td>12</td>\n",
       "      <td>57</td>\n",
       "      <td>GTGTGTGTGTGTGTGCTGTACTGTGCTGTGCTATGCATGTATTATG...</td>\n",
       "      <td>12</td>\n",
       "      <td>54</td>\n",
       "      <td>GTGTGTGCTGTGTGTACTGTGCTGTTCTATGCAGTGTGGTATGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>832657</td>\n",
       "      <td>832711</td>\n",
       "      <td>('ATGGACATGGGCGTAAGCAGGTTTG', 'ATGGATGGTCATGGG...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>54</td>\n",
       "      <td>CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...</td>\n",
       "      <td>2</td>\n",
       "      <td>54</td>\n",
       "      <td>...</td>\n",
       "      <td>CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...</td>\n",
       "      <td>2</td>\n",
       "      <td>54</td>\n",
       "      <td>CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...</td>\n",
       "      <td>2</td>\n",
       "      <td>54</td>\n",
       "      <td>CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...</td>\n",
       "      <td>2</td>\n",
       "      <td>54</td>\n",
       "      <td>CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>832737</td>\n",
       "      <td>832781</td>\n",
       "      <td>('GTTTT',)</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>45</td>\n",
       "      <td>GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT</td>\n",
       "      <td>9</td>\n",
       "      <td>45</td>\n",
       "      <td>...</td>\n",
       "      <td>GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT</td>\n",
       "      <td>9</td>\n",
       "      <td>45</td>\n",
       "      <td>GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT</td>\n",
       "      <td>9</td>\n",
       "      <td>45</td>\n",
       "      <td>GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT</td>\n",
       "      <td>9</td>\n",
       "      <td>45</td>\n",
       "      <td>GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778958</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138094452</td>\n",
       "      <td>138094474</td>\n",
       "      <td>('AAA', 'AAG')</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>AAAAAAAAGAAGAAGAAGAAGAAGAAGAAA</td>\n",
       "      <td>7</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>AAAAAAAAGAAGAAGAAGAAA</td>\n",
       "      <td>7</td>\n",
       "      <td>21</td>\n",
       "      <td>AAAAAAAAGAAGAAGAAGAAA</td>\n",
       "      <td>7</td>\n",
       "      <td>21</td>\n",
       "      <td>AAAAAAAAGAAGAAGAAGAAA</td>\n",
       "      <td>7</td>\n",
       "      <td>21</td>\n",
       "      <td>AAAAAAAAGAAGAAGAAGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778959</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138097462</td>\n",
       "      <td>138097499</td>\n",
       "      <td>('CTCA', 'CTCG', 'CTTG', 'TTC', 'TTTA')</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>39</td>\n",
       "      <td>CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA</td>\n",
       "      <td>10</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA</td>\n",
       "      <td>10</td>\n",
       "      <td>39</td>\n",
       "      <td>CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA</td>\n",
       "      <td>10</td>\n",
       "      <td>39</td>\n",
       "      <td>CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA</td>\n",
       "      <td>10</td>\n",
       "      <td>39</td>\n",
       "      <td>CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778960</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138102608</td>\n",
       "      <td>138102664</td>\n",
       "      <td>('CCCTAC', 'CCCTCC', 'CCTCC', 'CGCTC', 'CGCTCC...</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>53</td>\n",
       "      <td>CCTCCCTGCCCCTGCCCCTGCCCCTGCCCCGCTCTCCTCCTGCTGC...</td>\n",
       "      <td>9</td>\n",
       "      <td>53</td>\n",
       "      <td>...</td>\n",
       "      <td>CCTCCCTGCCCCTGCCCCTGCCCCTGCCCCGCTCTCCTCCTGCTGC...</td>\n",
       "      <td>9</td>\n",
       "      <td>53</td>\n",
       "      <td>CCTCCCTGCCCCTGCCCCTGCCCCTGCCCCGCTCTCCTCCTGCTGC...</td>\n",
       "      <td>9</td>\n",
       "      <td>53</td>\n",
       "      <td>CCTCCCTGCCCCTGCCCCTGCCCCTGCCCCGCTCTCCTCCTGCTGC...</td>\n",
       "      <td>9</td>\n",
       "      <td>54</td>\n",
       "      <td>CCCTCCCTGCCCCTGCCCCGCTCCCCTCCCGCTCCCCTCCTGCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778961</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138106311</td>\n",
       "      <td>138106353</td>\n",
       "      <td>('CCCACCC', 'CCCAGC', 'CCCATC', 'CCCTGC', 'GTC...</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>42</td>\n",
       "      <td>CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTG</td>\n",
       "      <td>7</td>\n",
       "      <td>42</td>\n",
       "      <td>...</td>\n",
       "      <td>CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTG</td>\n",
       "      <td>7</td>\n",
       "      <td>42</td>\n",
       "      <td>CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTG</td>\n",
       "      <td>7</td>\n",
       "      <td>42</td>\n",
       "      <td>CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTG</td>\n",
       "      <td>7</td>\n",
       "      <td>43</td>\n",
       "      <td>CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778962</th>\n",
       "      <td>chr9</td>\n",
       "      <td>138107215</td>\n",
       "      <td>138107268</td>\n",
       "      <td>('TTAT',)</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>52</td>\n",
       "      <td>TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTT...</td>\n",
       "      <td>10</td>\n",
       "      <td>40</td>\n",
       "      <td>...</td>\n",
       "      <td>TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTT...</td>\n",
       "      <td>15</td>\n",
       "      <td>60</td>\n",
       "      <td>TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTT...</td>\n",
       "      <td>12</td>\n",
       "      <td>48</td>\n",
       "      <td>TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTAT</td>\n",
       "      <td>12</td>\n",
       "      <td>48</td>\n",
       "      <td>TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTAT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>778963 rows × 215 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      start        end  \\\n",
       "0       chr1     819914     823663   \n",
       "1       chr1     827584     827629   \n",
       "2       chr1     829414     829478   \n",
       "3       chr1     832657     832711   \n",
       "4       chr1     832737     832781   \n",
       "...      ...        ...        ...   \n",
       "778958  chr9  138094452  138094474   \n",
       "778959  chr9  138097462  138097499   \n",
       "778960  chr9  138102608  138102664   \n",
       "778961  chr9  138106311  138106353   \n",
       "778962  chr9  138107215  138107268   \n",
       "\n",
       "                                                   motifs  CN_ref  \\\n",
       "0       ('ACTACCCTATACAGCAGA', 'AGGTCCACATTGTCCAGC', '...     212   \n",
       "1       ('GCCGGGGCGGGGT', 'GCGGGGCCGGGCCGG', 'GGCGGGCT...       3   \n",
       "2       ('ATGCA', 'ATGCAT', 'GGGT', 'GGT', 'GTACT', 'G...      12   \n",
       "3       ('ATGGACATGGGCGTAAGCAGGTTTG', 'ATGGATGGTCATGGG...       2   \n",
       "4                                              ('GTTTT',)       9   \n",
       "...                                                   ...     ...   \n",
       "778958                                     ('AAA', 'AAG')       7   \n",
       "778959            ('CTCA', 'CTCG', 'CTTG', 'TTC', 'TTTA')      10   \n",
       "778960  ('CCCTAC', 'CCCTCC', 'CCTCC', 'CGCTC', 'CGCTCC...       9   \n",
       "778961  ('CCCACCC', 'CCCAGC', 'CCCATC', 'CCCTGC', 'GTC...       7   \n",
       "778962                                          ('TTAT',)      13   \n",
       "\n",
       "        CN_NA18939_h2  len_NA18939_h2  \\\n",
       "0                 223            3910   \n",
       "1                   3              43   \n",
       "2                  12              57   \n",
       "3                   2              54   \n",
       "4                   9              45   \n",
       "...               ...             ...   \n",
       "778958             10              30   \n",
       "778959             10              39   \n",
       "778960              9              53   \n",
       "778961              7              42   \n",
       "778962             13              52   \n",
       "\n",
       "                                           seq_NA18939_h2  CN_NA18534_h2  \\\n",
       "0       CTGTCGAGCAGATCCACCCTGTCTATACTACCTGACTCTCCAGAAG...            223   \n",
       "1             GCGGGGCCGGGCCGGGCGGGGCCGGGCCGGGCCGGGGCGGGGT              3   \n",
       "2       GTGTGTGTGTGTGTGCTGTACTGTGCTGTGCTATGCATGTATTATG...             12   \n",
       "3       CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...              2   \n",
       "4           GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT              9   \n",
       "...                                                   ...            ...   \n",
       "778958                     AAAAAAAAGAAGAAGAAGAAGAAGAAGAAA              7   \n",
       "778959            CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA             10   \n",
       "778960  CCTCCCTGCCCCTGCCCCTGCCCCTGCCCCGCTCTCCTCCTGCTGC...              9   \n",
       "778961         CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTG              7   \n",
       "778962  TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTT...             10   \n",
       "\n",
       "        len_NA18534_h2  ...  \\\n",
       "0                 3910  ...   \n",
       "1                   43  ...   \n",
       "2                   57  ...   \n",
       "3                   54  ...   \n",
       "4                   45  ...   \n",
       "...                ...  ...   \n",
       "778958              21  ...   \n",
       "778959              39  ...   \n",
       "778960              53  ...   \n",
       "778961              42  ...   \n",
       "778962              40  ...   \n",
       "\n",
       "                                           seq_HG03125_h1  CN_NA12878_h2  \\\n",
       "0       CTGTCGAGCAGATCCACCCTGTCTATACTACCTGACTCTCCAGAAG...            223   \n",
       "1             GCGGGGCCGGGCCGGGCGGGGCCGGGCCGGGCCGGGGCGGGGT              3   \n",
       "2       GTGTGTGTGTGTGTGCTGTACTGTGCTGTGCTATGCATGTATTATG...             12   \n",
       "3       CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...              2   \n",
       "4                GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT              9   \n",
       "...                                                   ...            ...   \n",
       "778958                              AAAAAAAAGAAGAAGAAGAAA              7   \n",
       "778959            CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA             10   \n",
       "778960  CCTCCCTGCCCCTGCCCCTGCCCCTGCCCCGCTCTCCTCCTGCTGC...              9   \n",
       "778961         CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTG              7   \n",
       "778962  TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTT...             15   \n",
       "\n",
       "        len_NA12878_h2                                     seq_NA12878_h2  \\\n",
       "0                 3910  CTGTCGAGCAGATCCACCCTGTCTATACTACCTGACTCTCCAGAAG...   \n",
       "1                   43        GCGGGGCCGGGCCGGGCGGGGCCGGGCCGGGCCGGGGCGGGGT   \n",
       "2                   57  GTGTGTGTGTGTGTGCTGTACTGTGCTGTGCTATGCATGTATTATG...   \n",
       "3                   54  CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...   \n",
       "4                   45      GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT   \n",
       "...                ...                                                ...   \n",
       "778958              21                              AAAAAAAAGAAGAAGAAGAAA   \n",
       "778959              39            CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA   \n",
       "778960              53  CCTCCCTGCCCCTGCCCCTGCCCCTGCCCCGCTCTCCTCCTGCTGC...   \n",
       "778961              42         CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTG   \n",
       "778962              60  TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTT...   \n",
       "\n",
       "        CN_HG00733_h1  len_HG00733_h1  \\\n",
       "0                 223            3910   \n",
       "1                   3              43   \n",
       "2                  12              57   \n",
       "3                   2              54   \n",
       "4                   9              45   \n",
       "...               ...             ...   \n",
       "778958              7              21   \n",
       "778959             10              39   \n",
       "778960              9              53   \n",
       "778961              7              42   \n",
       "778962             12              48   \n",
       "\n",
       "                                           seq_HG00733_h1  CN_NA24385_h2  \\\n",
       "0       CTGTCGAGCAGATCCACCCTGTCTATACTACCTGACTCTCCAGAAG...            224   \n",
       "1             GCGGGGCCGGGCCGGGCGGGGCCGGGCCGGGCCGGGGCGGGGT              3   \n",
       "2       GTGTGTGTGTGTGTGCTGTACTGTGCTGTGCTATGCATGTATTATG...             12   \n",
       "3       CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...              2   \n",
       "4           GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT              9   \n",
       "...                                                   ...            ...   \n",
       "778958                              AAAAAAAAGAAGAAGAAGAAA              7   \n",
       "778959            CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA             10   \n",
       "778960  CCTCCCTGCCCCTGCCCCTGCCCCTGCCCCGCTCTCCTCCTGCTGC...              9   \n",
       "778961         CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTG              7   \n",
       "778962   TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTAT             12   \n",
       "\n",
       "        len_NA24385_h2                                     seq_NA24385_h2  \n",
       "0                 3919  CTGTCGAGCAGATCCACCCTGTCTATACTACCTGCCTCTCCAGAAG...  \n",
       "1                   43        GCGGGGCCGGGCCGGGCCGGGGCGGGGTGCGGGGCCGGGCCGG  \n",
       "2                   54  GTGTGTGCTGTGTGTACTGTGCTGTTCTATGCAGTGTGGTATGCAT...  \n",
       "3                   54  CATGGACATGGGCGTAAGCAGGTTTGATGGTCATGGGCATAGGCAG...  \n",
       "4                   45      GTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTTGTTTT  \n",
       "...                ...                                                ...  \n",
       "778958              21                              AAAAAAAAGAAGAAGAAGAAG  \n",
       "778959              39            CTCACTCGTTCCTCATTTACTTGCTCGCTCACTCACTCA  \n",
       "778960              54  CCCTCCCTGCCCCTGCCCCGCTCCCCTCCCGCTCCCCTCCTGCTGC...  \n",
       "778961              43        CCCTGCCCCAGCCCCACCCCCCTGCGTCTGCCCCATCTCCTGC  \n",
       "778962              48   TTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTAT  \n",
       "\n",
       "[778963 rows x 215 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "path = '/confidential/tGenVar/vntr/output_maryam/tools/tandemtwister-vis-maryam/scripts/output'\n",
    "plt.savefig(os.path.join(path, 'population', 'cum_allele_count.pdf'), format='pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computing sequence similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "asm_h1_file = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/hgsvc/TandemTwist/asm/cut_reads_NA24385_h1.tsv.fasta\"\n",
    "asm_h2_file = \"/confidential/tGenVar/vntr/output_maryam/tools/run_all_tools/output/hgsvc/TandemTwist/asm/cut_reads_NA24385_h2.tsv.fasta\"\n",
    "\n",
    "asm_h1 = parse_fasta.parse_fasta_reads_cut(fasta_file = asm_h1_file, seq_col_name = \"asm_h1_seq\")\n",
    "asm_h2 = parse_fasta.parse_fasta_reads_cut(fasta_file = asm_h2_file, seq_col_name = \"asm_h2_seq\")\n",
    "\n",
    "asm_df = pd.merge(asm_h1, asm_h2, on=['chrom', 'start', 'end'], how='inner')\n",
    "\n",
    "asm_h1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_df = asm_df[\n",
    "    (asm_df['asm_h1_seq'].isnull()) |\n",
    "    (asm_df['asm_h2_seq'].isnull())\n",
    "]\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "child_df_tandem_twister_compare = compare_with_asm.compare_sample_to_asm(child_df_tandem_twister.head(100000), asm_df)\n",
    "# memory allocation error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "subset  = child_df_tandem_twister_compare[\n",
    "    child_df_tandem_twister_compare['seq_sim'] < 0.55\n",
    "]\n",
    "\n",
    "# subset['motif_ids_H1', 'motif_ids_H2', 'motif_concat_h1', 'motif_concat_h2', '']\n",
    "\n",
    "filtered_df = child_df_tandem_twister_compare[\n",
    "    (child_df_tandem_twister_compare['motif_concat_h1'] != \"NA\") &\n",
    "    (child_df_tandem_twister_compare['motif_concat_h2'] != \"NA\")\n",
    "]\n",
    "filtered_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Pairs of dataframes and their corresponding output labels\n",
    "dataframe_pairs = [\n",
    "    (child_df_tandem_twister.head(100000), \"child_df_tandem_twister_compare\"),\n",
    "    #(child_df_tandem_twister_clr.head(10000), \"child_df_tandem_twister_clr_compare\"),\n",
    "    #(child_df_tandem_twister_ont.head(10000), \"child_df_tandem_twister_ont_compare\"),\n",
    "    (child_df_vamos.head(100000), \"child_df_vamos_compare\"),\n",
    "    #(child_df_vamos_clr.head(10000), \"child_df_vamos_clr_compare\"),\n",
    "    #(child_df_vamos_ont.head(10000), \"child_df_vamos_ont_compare\"),\n",
    "    (child_df_tandem_TRGT.head(100000), \"child_df_TRGT_compare\")\n",
    "]\n",
    "\n",
    "# Dictionary to store results\n",
    "results = {}\n",
    "\n",
    "# Process each pair of DataFrames one by one\n",
    "for df, label in dataframe_pairs:\n",
    "    try:\n",
    "        # Call the function with each pair of DataFrames\n",
    "        result = compare_with_asm.compare_sample_to_asm(df, asm_df)\n",
    "        mean_seq_sim = result['seq_sim'].dropna().mean()\n",
    "        mean_len_sim = result['len_sim'].dropna().mean()\n",
    "        print(f\"{label}, mean seq_sim = {mean_seq_sim}, mean len_sim = {mean_len_sim}\")\n",
    "        # Store the result using the label as the key\n",
    "        results[label] = result\n",
    "    except Exception as e:\n",
    "        print(f'Function call raised an exception for {label}: {e}')\n",
    "\n",
    "# # Using ProcessPoolExecutor for parallel execution across multiple cores\n",
    "# with concurrent.futures.ProcessPoolExecutor() as executor:\n",
    "#     # Map the function over the dataframe pairs\n",
    "#     future_to_label = {\n",
    "#         executor.submit(compare_with_asm.compare_sample_to_asm, df, asm_df): label\n",
    "#         for df, label in dataframe_pairs\n",
    "#     }\n",
    "\n",
    "#     # Collect results as they complete\n",
    "#     for future in concurrent.futures.as_completed(future_to_label):\n",
    "#         label = future_to_label[future]\n",
    "#         try:\n",
    "#             result = future.result()\n",
    "#             results[label] = result\n",
    "#         except Exception as e:\n",
    "#             print(f'Function call raised an exception for {label}: {e}')\n",
    "\n",
    "\n",
    "# print('compare tandemtwister ccs with assembly')\n",
    "# child_df_tandem_twister_compare = compare_with_asm.compare_sample_to_asm(child_df_tandem_twister.head(100000), asm_df)\n",
    "# pd.set_option('display.max_colwidth', None)\n",
    "# child_df_tandem_twister_compare[['chrom', 'start',\t'end', 'motif_concat_h1', 'motif_concat_h2', 'asm_h1_seq', 'asm_h2_seq', 'seq_sim', 'len_sim', 'assigned_haplotype']]\n",
    "# tandemtwist_ccs_merge_asm.to_csv('tandemtwist_ccs_compare_asm.csv', index=False)\n",
    "# 882 s --- one df size 1.2 M\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# heatmap hex plots \n",
    "tantwist = sns.jointplot(data = tandemtwist_ccs_merge_asm, x='len_sim', y='seq_sim', kind=\"hex\", color=\"#4CB391\")\n",
    "tantwist.fig.suptitle(\"TandemTwister\")\n",
    "\n",
    "# plt2 --- vamos ccs\n",
    "vamos_plt = sns.jointplot(data = vamos_ccs_merge_asm, x='len_sim', y='seq_sim', kind=\"hex\", color=\"#4CB391\")\n",
    "vamos_plt.fig.suptitle(\"Vamos\")\n",
    "\n",
    "# plt3 --- trgt ccs\n",
    "vamos_plt = sns.jointplot(data = trgt_ccs_merge_asm, x='len_sim', y='seq_sim', kind=\"hex\", color=\"#4CB391\")\n",
    "vamos_plt.fig.suptitle(\"TRGT\")\n",
    "\n",
    "\n",
    "# statistics:\n",
    "mean_seq_sim_tantwist = tandemtwist_ccs_merge_asm['seq_sim'].dropna().mean()\n",
    "median_seq_sim_tantwist = tandemtwist_ccs_merge_asm['seq_sim'].dropna().median()\n",
    "mean_len_sim_tantwist = tandemtwist_ccs_merge_asm['len_sim'].dropna().mean()\n",
    "median_len_sim_tantwist = tandemtwist_ccs_merge_asm['len_sim'].dropna().median()\n",
    "\n",
    "mean_seq_sim_vamos = vamos_ccs_merge_asm['seq_sim'].dropna().mean()\n",
    "median_seq_sim_vamos = vamos_ccs_merge_asm['seq_sim'].dropna().median()\n",
    "mean_len_sim_vamos = vamos_ccs_merge_asm['len_sim'].dropna().mean()\n",
    "median_len_sim_vamos = vamos_ccs_merge_asm['len_sim'].dropna().median()\n",
    "\n",
    "\n",
    "mean_seq_sim_trgt = trgt_ccs_merge_asm['seq_sim'].dropna().mean()\n",
    "median_seq_sim_trgt = trgt_ccs_merge_asm['seq_sim'].dropna().median()\n",
    "mean_len_sim_trgt = trgt_ccs_merge_asm['len_sim'].dropna().mean()\n",
    "median_len_sim_trgt = trgt_ccs_merge_asm['len_sim'].dropna().median()\n",
    "\n",
    "\n",
    "# Create a dictionary to structure your data\n",
    "data = {\n",
    "    'mean_seq_sim': [mean_seq_sim_tantwist, mean_seq_sim_vamos, mean_seq_sim_trgt],\n",
    "    'median_seq_sim': [median_seq_sim_tantwist, median_seq_sim_vamos, median_seq_sim_trgt],\n",
    "    'mean_len_sim': [mean_len_sim_tantwist, mean_len_sim_vamos, mean_len_sim_trgt],\n",
    "    'median_len_sim': [median_len_sim_tantwist, median_len_sim_vamos, median_len_sim_trgt]\n",
    "}\n",
    "\n",
    "# Define the index labels (row labels)\n",
    "index_labels = ['tantwist', 'vamos', 'trgt']\n",
    "\n",
    "# Create a DataFrame\n",
    "df = pd.DataFrame(data, index=index_labels)\n",
    "\n",
    "# Display the DataFrame\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('compare trgt ccs with assembly')\n",
    "trgt_ccs_merge_asm = compare_with_asm.compare_sample_to_asm(child_df_tandem_TRGT.head(100000), asm_df)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "trgt_ccs_merge_asm[['chrom', 'start',\t'end', 'motif_concat_h1', 'motif_concat_h2', 'asm_h1_seq', 'asm_h2_seq', 'seq_sim', 'len_sim', 'assigned_haplotype']]\n",
    "\n",
    "# trgt_ccs_merge_asm.to_csv('trgt_ccs_compare_asm.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('compare trgt ccs with assembly')\n",
    "vamos_ccs_merge_asm = compare_with_asm.compare_sample_to_asm(child_df_vamos.head(100000), asm_df)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "vamos_ccs_merge_asm[['chrom', 'start',\t'end', 'motif_concat_h1', 'motif_concat_h2', 'asm_h1_seq', 'asm_h2_seq', 'seq_sim', 'len_sim', 'assigned_haplotype']]\n",
    "\n",
    "vamos_ccs_merge_asm.to_csv('vamos_ccs_compare_asm.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = plot_heatmap_with_histograms(tandemtwist_ccs_merge_asm.head(10000))\n",
    "plot.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tandemtwist_compare_asm[['MAE_motif_count', 'relative_MAE_motif_count', 'ambiguous_motifs']] = tandemtwist_compare_asm.apply(compare_counts_with_asm, axis=1, result_type='expand')\n",
    "tandemtwist_clr_compare_asm[['MAE_motif_count', 'relative_MAE_motif_count', 'ambiguous_motifs']] = tandemtwist_clr_compare_asm.apply(compare_counts_with_asm, axis=1, result_type='expand')\n",
    "tandemtwist_ont_compare_asm[['MAE_motif_count', 'relative_MAE_motif_count', 'ambiguous_motifs']] = tandemtwist_ont_compare_asm.apply(compare_counts_with_asm, axis=1, result_type='expand')\n",
    "\n",
    "tandemtwist_clr_compare_asm\n",
    "\n",
    "# true_count = tandemtwist_compare_asm['ambiguous_motifs'].sum()\n",
    "# print(f\"Number of 'all ambiguous_motifs' regions: {true_count}\") # 207324"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vamos_compare_asm[['MAE_motif_count', 'relative_MAE_motif_count', 'ambiguous_motifs']] = vamos_compare_asm.apply(compare_counts_with_asm, axis=1, result_type='expand')\n",
    "vamos_clr_compare_asm[['MAE_motif_count', 'relative_MAE_motif_count', 'ambiguous_motifs']] = vamos_clr_compare_asm.apply(compare_counts_with_asm, axis=1, result_type='expand')\n",
    "vamos_ont_compare_asm[['MAE_motif_count', 'relative_MAE_motif_count', 'ambiguous_motifs']] = vamos_ont_compare_asm.apply(compare_counts_with_asm, axis=1, result_type='expand')\n",
    "\n",
    "vamos_clr_compare_asm\n",
    "\n",
    "# true_count = vamos_compare_asm['ambiguous_motifs'].sum()\n",
    "# print(f\"Number of 'all ambiguous_motifs' regions: {true_count}\") # 210781"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trgt_compare_asm[['MAE_motif_count', 'relative_MAE_motif_count', 'ambiguous_motifs']] = trgt_compare_asm.apply(compare_counts_with_asm, axis=1, result_type='expand')\n",
    "\n",
    "trgt_compare_asm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Merge DataFrames on 'chrom', 'start', 'end'\n",
    "\n",
    "\n",
    "# First, merge the first two DataFrames\n",
    "merged_df_1 = pd.merge(\n",
    "    tandemtwist_compare_asm[['chrom', 'start', 'end', 'MAE_motif_count', 'relative_MAE_motif_count']],\n",
    "    vamos_compare_asm[['chrom', 'start', 'end', 'MAE_motif_count', 'relative_MAE_motif_count']],\n",
    "    on=['chrom', 'start', 'end'],\n",
    "    suffixes=('_tandemtwist', '_vamos')\n",
    ")\n",
    "\n",
    "# Then, merge the result with the third DataFrame\n",
    "merged_df = pd.merge(\n",
    "    merged_df_1,\n",
    "    trgt_compare_asm[['chrom', 'start', 'end', 'MAE_motif_count', 'relative_MAE_motif_count']],\n",
    "    on=['chrom', 'start', 'end'],\n",
    "    suffixes=('', '_trgt')  # No suffix for the first, and '_trgt' for the second\n",
    ")\n",
    "\n",
    "# Rename the columns in merged_df to include the suffix '_trgt'\n",
    "merged_df.rename(columns={\n",
    "    'MAE_motif_count': 'MAE_motif_count_trgt',\n",
    "    'relative_MAE_motif_count': 'relative_MAE_motif_count_trgt'\n",
    "}, inplace=True)\n",
    "\n",
    "\n",
    "# # Calculate mean and median values of relative_MAE for each tool\n",
    "# mean_relative_MAE_tandemtwist = merged_df['relative_MAE_motif_count_tandemtwist'].dropna().mean()\n",
    "# mean_relative_MAE_vamos = merged_df['relative_MAE_motif_count_vamos'].dropna().mean()\n",
    "# mean_relative_MAE_trgt = merged_df['relative_MAE_motif_count_trgt'].dropna().mean()\n",
    "# median_relative_MAE_tandemtwist = merged_df['relative_MAE_motif_count_tandemtwist'].dropna().median()\n",
    "# median_relative_MAE_vamos = merged_df['relative_MAE_motif_count_vamos'].dropna().median()\n",
    "# median_relative_MAE_trgt = merged_df['relative_MAE_motif_count_trgt'].dropna().median()\n",
    "\n",
    "# print(f\"Mean Relative MAE (TandemTwist): {mean_relative_MAE_tandemtwist}\")\n",
    "# print(f\"Mean Relative MAE (VAMOS): {mean_relative_MAE_vamos}\")\n",
    "# print(f\"Mean Relative MAE (TRGT): {mean_relative_MAE_trgt}\")\n",
    "# print(f\"Median Relative MAE (TandemTwist): {median_relative_MAE_tandemtwist}\")\n",
    "# print(f\"Median Relative MAE (VAMOS): {median_relative_MAE_vamos}\")\n",
    "# print(f\"Median Relative MAE (TRGT): {median_relative_MAE_trgt}\")\n",
    "\n",
    "# # Trim the top 5% of relative_MAEs for each tool\n",
    "tandemtwist_trimmed = merged_df['MAE_motif_count_tandemtwist'].dropna() # merged_df['relative_MAE_motif_count_tandemtwist'].dropna()\n",
    "vamos_trimmed = merged_df['MAE_motif_count_vamos'].dropna() # merged_df['relative_MAE_motif_count_vamos'].dropna()\n",
    "trgt_trimmed = merged_df['MAE_motif_count_trgt'].dropna() # merged_df['relative_MAE_motif_count_trgt'].dropna()\n",
    "\n",
    "tandemtwist_trimmed = tandemtwist_trimmed[tandemtwist_trimmed <= tandemtwist_trimmed.quantile(0.95)]\n",
    "vamos_trimmed = vamos_trimmed[vamos_trimmed <= vamos_trimmed.quantile(0.95)]\n",
    "trgt_trimmed = trgt_trimmed[trgt_trimmed <= trgt_trimmed.quantile(0.95)]\n",
    "\n",
    "\n",
    "# Calculate mean and median values of MAE for each tool\n",
    "mean_MAE_tandemtwist = merged_df['MAE_motif_count_tandemtwist'].dropna().mean()\n",
    "mean_MAE_vamos = merged_df['MAE_motif_count_vamos'].dropna().mean()\n",
    "mean_MAE_trgt = merged_df['MAE_motif_count_trgt'].dropna().mean()\n",
    "# median_MAE_tandemtwist = merged_df['MAE_motif_count_tandemtwist'].dropna().median()\n",
    "# median_MAE_vamos = merged_df['MAE_motif_count_vamos'].dropna().median()\n",
    "# median_MAE_trgt = merged_df['MAE_motif_count_trgt'].dropna().median()\n",
    "\n",
    "print(f\"Mean MAE (TandemTwist): {mean_MAE_tandemtwist}\")\n",
    "print(f\"Mean MAE (VAMOS): {mean_MAE_vamos}\")\n",
    "print(f\"Mean MAE (TRGT): {mean_MAE_trgt}\")\n",
    "\n",
    "# Prepare data for plotting\n",
    "trimmed_data = pd.DataFrame({\n",
    "    'Tool': ['TandemTwist'] * len(tandemtwist_trimmed) + ['VAMOS'] * len(vamos_trimmed) + ['TRGT'] * len(trgt_trimmed),\n",
    "    'MAE': pd.concat([tandemtwist_trimmed, vamos_trimmed, trgt_trimmed])\n",
    "})\n",
    "\n",
    "# Plot (violin)\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.violinplot(data=trimmed_data, x='Tool', y='MAE')\n",
    "plt.title(\"Trimmed Violin Plot of MAE for Different Tools (Top 5% Removed)\")\n",
    "plt.show()\n",
    "\n",
    "# # Create box plots\n",
    "# plt.figure(figsize=(10, 6))\n",
    "# plt.boxplot([trimmed_data[trimmed_data['Tool'] == 'TandemTwist']['MAE'],\n",
    "#              trimmed_data[trimmed_data['Tool'] == 'VAMOS']['MAE']],\n",
    "#             labels=['TandemTwist', 'VAMOS'])\n",
    "\n",
    "# plt.ylabel('MAE')\n",
    "# plt.title('Box Plot of MAE for Different Tools (Top 5% Removed)')\n",
    "# plt.grid(axis='y')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df_clr = pd.merge(\n",
    "    tandemtwist_clr_compare_asm[['chrom', 'start', 'end', 'MAE_motif_count', 'relative_MAE_motif_count']],\n",
    "    vamos_clr_compare_asm[['chrom', 'start', 'end', 'MAE_motif_count', 'relative_MAE_motif_count']],\n",
    "    on=['chrom', 'start', 'end'],\n",
    "    suffixes=('_tandemtwist', '_vamos')\n",
    ")\n",
    "\n",
    "merged_df_ont = pd.merge(\n",
    "    tandemtwist_ont_compare_asm[['chrom', 'start', 'end', 'MAE_motif_count', 'relative_MAE_motif_count']],\n",
    "    vamos_ont_compare_asm[['chrom', 'start', 'end', 'MAE_motif_count', 'relative_MAE_motif_count']],\n",
    "    on=['chrom', 'start', 'end'],\n",
    "    suffixes=('_tandemtwist', '_vamos')\n",
    ")\n",
    "\n",
    "tandemtwist_trimmed_clr = merged_df_clr['MAE_motif_count_tandemtwist'].dropna() \n",
    "vamos_trimmed_clr = merged_df_clr['MAE_motif_count_vamos'].dropna()\n",
    "tandemtwist_trimmed_clr = tandemtwist_trimmed_clr[tandemtwist_trimmed_clr <= tandemtwist_trimmed_clr.quantile(0.95)]\n",
    "vamos_trimmed_clr = vamos_trimmed_clr[vamos_trimmed_clr <= vamos_trimmed_clr.quantile(0.95)]\n",
    "\n",
    "tandemtwist_trimmed_ont = merged_df_ont['MAE_motif_count_tandemtwist'].dropna()\n",
    "vamos_trimmed_ont = merged_df_ont['MAE_motif_count_vamos'].dropna()\n",
    "tandemtwist_trimmed_ont = tandemtwist_trimmed_ont[tandemtwist_trimmed_ont <= tandemtwist_trimmed_ont.quantile(0.95)]\n",
    "vamos_trimmed_ont = vamos_trimmed_ont[vamos_trimmed_ont <= vamos_trimmed_ont.quantile(0.95)]\n",
    "\n",
    "mean_MAE_tandemtwist_clr = merged_df_clr['MAE_motif_count_tandemtwist'].dropna().mean()\n",
    "mean_MAE_vamos_clr = merged_df_clr['MAE_motif_count_vamos'].dropna().mean()\n",
    "print(f\"Mean MAE CLR (TandemTwist): {mean_MAE_tandemtwist_clr}\")\n",
    "print(f\"Mean MAE CLR (VAMOS): {mean_MAE_vamos_clr}\")\n",
    "\n",
    "mean_MAE_tandemtwist_ont = merged_df_ont['MAE_motif_count_tandemtwist'].dropna().mean()\n",
    "mean_MAE_vamos_ont = merged_df_ont['MAE_motif_count_vamos'].dropna().mean()\n",
    "print(f\"Mean MAE ONT (TandemTwist): {mean_MAE_tandemtwist_ont}\")\n",
    "print(f\"Mean MAE ONT (VAMOS): {mean_MAE_vamos_ont}\")\n",
    "\n",
    "# Prepare data for plotting\n",
    "trimmed_data = pd.DataFrame({\n",
    "    'Tool': ['TandemTwist'] * len(tandemtwist_trimmed_clr) + ['VAMOS'] * len(vamos_trimmed_clr) + ['TandemTwist'] * len(tandemtwist_trimmed_ont) + ['VAMOS'] * len(vamos_trimmed_ont),\n",
    "    'Tech': ['CLR'] * (len(tandemtwist_trimmed_clr)+len(vamos_trimmed_clr)) + ['ONT'] * (len(tandemtwist_trimmed_ont)+len(vamos_trimmed_ont)),\n",
    "    'MAE': pd.concat([tandemtwist_trimmed_clr, vamos_trimmed_clr, tandemtwist_trimmed_ont, vamos_trimmed_ont])\n",
    "})\n",
    "\n",
    "# Plot (violin)\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.violinplot(data=trimmed_data, x='Tech', hue='Tool', y='MAE')\n",
    "plt.title(\"Trimmed Violin Plot of MAE for Different Tools (Top 5% Removed)\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the motif file\n",
    "motif_file = \"/confidential/tGenVar/Lion/TandemTwist/vamos.motif.hg38.v2.1.e0.1.tsv\"\n",
    "# read the first 4 columns of the motif file\n",
    "motif_df = pd.read_csv(motif_file, sep='\\t', usecols=[0, 1, 2, 3], header=None)\n",
    "# create a header \n",
    "motif_df.columns = ['chrom', 'start', 'end', 'motifs']\n",
    "# merge the child results of the three tools with the motif file\n",
    "df = motif_df.merge(child_df_tandem_twister, on=['chrom', 'start', 'end'], how='left', suffixes=('_motif', '_tandem_twister'))\n",
    "df = df.merge(child_df_tandem_TRGT, on=['chrom', 'start', 'end',], how='left', suffixes=('_tandem_twister', '_TRGT'))\n",
    "df = df.merge(child_df_vamos, on=['chrom', 'start', 'end'], how='left', suffixes=('_TRGT', '_vamos'))\n",
    "\n",
    "copynumbers = df[['chrom','start','end','motifs_motif','CN_H1_tandem_twister','CN_H2_tandem_twister','CN_H1_TRGT','CN_H2_TRGT','CN_H1','CN_H2' ]]\n",
    "copynumbers.columns = ['chrom','start','end','motifs','CN_H1_tandem_twister','CN_H2_tandem_twister','CN_H1_TRGT','CN_H2_TRGT','CN_H1_vamos','CN_H2_vamos']\n",
    "# make the CN columns as integers\n",
    "# fill the nan values with 0\n",
    "copynumbers = copynumbers.fillna(0)\n",
    "\n",
    "copynumbers['CN_H1_tandem_twister'] = copynumbers['CN_H1_tandem_twister'].astype(int)\n",
    "copynumbers['CN_H2_tandem_twister'] = copynumbers['CN_H2_tandem_twister'].astype(int)\n",
    "copynumbers['CN_H1_TRGT'] = copynumbers['CN_H1_TRGT'].astype(int)\n",
    "copynumbers['CN_H2_TRGT'] = copynumbers['CN_H2_TRGT'].astype(int)\n",
    "copynumbers['CN_H1_vamos'] = copynumbers['CN_H1_vamos'].astype(int)\n",
    "copynumbers['CN_H2_vamos'] = copynumbers['CN_H2_vamos'].astype(int)\n",
    "copynumbers['agreement'] = copynumbers.apply(check_agreement, axis=1)\n",
    "\n",
    "\n",
    "\n",
    "copynumbers['tandem_twister_disagree'] = copynumbers['agreement'] == 'tandem_twister_disagree'\n",
    "copynumbers['TRGT_disagree'] = copynumbers['agreement'] == 'TRGT_disagree'\n",
    "copynumbers['vamos_disagree'] = copynumbers['agreement'] == 'vamos_disagree'\n",
    "copynumbers['all_disagree'] = copynumbers['agreement'] == 'all_disagree'\n",
    "copynumbers['all_agree'] = copynumbers['agreement'] == 'all_agree'\n",
    "\n",
    "# Convert to integer binary representation (0 or 1)\n",
    "binary_columns = copynumbers[['tandem_twister_disagree', 'TRGT_disagree', 'vamos_disagree', 'all_disagree', 'all_agree']].astype(int)\n",
    "\n",
    "# Group the data by combinations of binary columns and count occurrences\n",
    "counts = binary_columns.groupby(\n",
    "    ['tandem_twister_disagree', 'TRGT_disagree', 'vamos_disagree', 'all_disagree', 'all_agree']\n",
    ").size()\n",
    "\n",
    "# Reset the index to convert it to a MultiIndex, which is required for UpSet\n",
    "counts = counts.reset_index()\n",
    "\n",
    "# Convert columns into a MultiIndex for the UpSet plot\n",
    "counts = counts.set_index(['tandem_twister_disagree', 'TRGT_disagree', 'vamos_disagree', 'all_disagree', 'all_agree'])[0]\n",
    "\n",
    "# Create the UpSet plot\n",
    "upset = UpSet(counts, orientation='horizontal')\n",
    "upset.plot()\n",
    "\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 200)\n",
    "copynumbers.head(200)\n",
    "# make an upset plot from the column all_agree\n",
    "copynumbers[copynumbers['tandem_twister_disagree'] == 1].head(200)\n",
    "# check this region for tandemTwister\t\n",
    "# chr1:104161-104196\t\n",
    "# chr1\t109565\t109614 # in this region tandemtwister is right and vamos, TRGT are wrong\n",
    "# check these : chr1\t178283\t178336,  chr1\t933011\t933048\n",
    "# chr1\t414630\t414675\tthe region is clearly hetrotzygous but vamos and TRGT reports it as homozygous\n",
    "# chr1\t640899\t640928 homopolymer region, TRGT and Vamos are wrong in this region\n",
    "# chr1\t1845825\t1845872  # this is an interesting region, tandemtwister is right and vamos, TRGT are wrong, it is a hetrotzygous region\n",
    "# interesting region : chr1:432,156-432,331\n",
    "# look at this region in TandemTwister chr1\t9311161\t9311178 we say it's hetrotzygous but TRGT and Vamos say it's homozygous which is true \n",
    "# chr1\t1604266\t1604292 this region is genotyped correctly by tandemtwister but not by others\n",
    "# very differen results :chr1\t7551284\t7551372\t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_number_of_motifs_found_in_region(row):\n",
    "    # get the unique motifs in the region\n",
    "    if pd.isna(row['motif_ids_H1']) or pd.isna(row['motif_ids_H2']):\n",
    "        return 0\n",
    "    \n",
    "    return max(len(set(row['motif_ids_H1'])), len(set(row['motif_ids_H2'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "child_df_tandem_twister['number_of_motifs_found'] = child_df_tandem_twister.apply(get_number_of_motifs_found_in_region, axis=1)\n",
    "child_df_tandem_TRGT['number_of_motifs_found'] = child_df_tandem_TRGT.apply(get_number_of_motifs_found_in_region, axis=1)\n",
    "child_df_vamos['number_of_motifs_found'] = child_df_vamos.apply(get_number_of_motifs_found_in_region, axis=1)\n",
    "\n",
    "child_df_tandem_twister['index'] = child_df_tandem_twister.index\n",
    "child_df_tandem_TRGT['index'] = child_df_tandem_TRGT.index\n",
    "child_df_vamos['index'] = child_df_vamos.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Create the figure and axis\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Plot the data with different markers and transparency for better distinction\n",
    "ax.scatter(child_df_tandem_twister['index'], child_df_tandem_twister['number_of_motifs_found'], \n",
    "           label='Tandem Twister', marker='o', color='blue', alpha=0.7)\n",
    "ax.scatter(child_df_vamos['index'], child_df_vamos['number_of_motifs_found'], \n",
    "           label='Vamos', marker='x', color='green', alpha=0.7)\n",
    "ax.scatter(child_df_tandem_TRGT['index'], child_df_tandem_TRGT['number_of_motifs_found'], \n",
    "           label='TRGT', marker='^', color='red', alpha=0.7)\n",
    "\n",
    "# Set labels and title\n",
    "ax.set_xlabel('Region', fontsize=12)\n",
    "ax.set_ylabel('Number of Motifs Found', fontsize=12)\n",
    "ax.set_title('Comparison of Motif Finding Tools', fontsize=14)\n",
    "\n",
    "# Add gridlines for easier reading of points\n",
    "ax.grid(True)\n",
    "\n",
    "# Add a legend in the upper right corner\n",
    "ax.legend(loc='upper right', fontsize=10)\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "tandemtwister = get_mendelian_consistency(child_df_tandem_twister, father_df_tandem_twister, mother_df_tandem_twister, 'tandem_twister', child_sex)\n",
    "#print_mendalian_consistency_per_motif_size(tandemtwister)\n",
    "print (\"number of regions in tandemtwister \", tandemtwister.shape[0])   \n",
    "print(\"percentage of mendelian consistent regions for tandemtwister \", tandemtwister['mendelian_consistent_0'].sum()/tandemtwister.shape[0])\n",
    "print(\"percentage of mendelian consistent off by one regions for tandemtwister \", tandemtwister['mendelian_consistent_1'].sum()/tandemtwister.shape[0])\n",
    "print(\"percentage of mendelian consistent regions for tandemtwister with motif size > 1 \", tandemtwister[tandemtwister['max_motif_size'] > 1]['mendelian_consistent_0'].sum()/tandemtwister[tandemtwister['max_motif_size'] > 1].shape[0])\n",
    "print(\"percentage of mendelian consistent off by one regions for tandemtwister with motif size > 1 \", tandemtwister[tandemtwister['max_motif_size'] > 1]['mendelian_consistent_1'].sum()/tandemtwister[tandemtwister['max_motif_size'] > 1].shape[0])\n",
    "print(\"recall for tandemtwister \", tandemtwister.shape[0]/1243954)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print (\"number of regions in tandemtwister \", tandemtwister.shape[0])   \n",
    "print(\"percentage of mendelian consistent regions for tandemtwister \", tandemtwister['mendelian_consistent_0'].sum()/tandemtwister.shape[0])\n",
    "print(\"percentage of mendelian consistent regions for tandemtwister with motif size > 1 \", tandemtwister[tandemtwister['max_motif_size'] > 1]['mendelian_consistent_0'].sum()/tandemtwister[tandemtwister['max_motif_size'] > 1].shape[0])\n",
    "print(\"recall for tandemtwister \", tandemtwister.shape[0]/1243954)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tandemtwister[tandemtwister[\"mendelian_consistent_0\"] == True][[\"CN_H1_child\", \"CN_H2_child\", \"CN_H1_father\", \"CN_H2_father\", \"CN_H1_mother\", \"CN_H2_mother\", \"mendelian_consistent_0\", \"chrom\", \"start\", \"end\", \"motifs_child\", \"motif_ids_H1_child\", \"motif_ids_H2_child\", \"motif_ids_H1_father\", \"motif_ids_H2_father\", \"motif_ids_H1_mother\", \"motif_ids_H2_mother\"]]\n",
    "# count the number of regions where all CNs are 0\n",
    "tandemtwister[(tandemtwister[\"CN_H1_child\"] == 0) & (tandemtwister[\"CN_H2_child\"] == 0) & (tandemtwister[\"CN_H1_father\"] == 0) & (tandemtwister[\"CN_H2_father\"] == 0) & (tandemtwister[\"CN_H1_mother\"] == 0) & (tandemtwister[\"CN_H2_mother\"] == 0)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tandemtwister[tandemtwister[\"max_motif_size\"] ==1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TRGT = get_mendelian_consistency(child_df_tandem_TRGT, father_df_tandem_TRGT, mother_df_tandem_TRGT, 'TRGT',child_sex)\n",
    "#print_mendalian_consistency_per_motif_size(TRGT)\n",
    "print (\"number of  regions in TRGT \", TRGT.shape[0])\n",
    "print(\"Number of mendelian consistent regions for TRGT \", TRGT['mendelian_consistent_0'].sum()/TRGT.shape[0])\n",
    "print(\"Number of mendelian consistent off by one regions for TRGT \", TRGT['mendelian_consistent_1'].sum()/TRGT.shape[0])\n",
    "print(\"Number of mendelian consistent regions for TRGT with motif size > 1 \", TRGT[TRGT['max_motif_size'] > 1]['mendelian_consistent_0'].sum()/TRGT[TRGT['max_motif_size'] > 1].shape[0])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vamos = get_mendelian_consistency(child_df_vamos, father_df_vamos, mother_df_vamos, 'vamos',child_sex)\n",
    "#print_mendalian_consistency_per_motif_size(vamos)\n",
    "print (\"number of regions in vamos \", vamos.shape[0])\n",
    "print(\"Number of mendelian consistent regions for vamos \", vamos['mendelian_consistent_0'].sum()/vamos.shape[0])\n",
    "print(\"Number of mendelian consistent off by one regions for vamos \", vamos['mendelian_consistent_1'].sum()/vamos.shape[0])\n",
    "print(\"Number of mendelian consistent regions for vamos with motif size > 1 \", vamos[vamos['max_motif_size'] > 1]['mendelian_consistent_0'].sum()/vamos[vamos['max_motif_size'] > 1].shape[0])\n",
    "print(\"Number of mendelian consistent regions off by one for vamos with motif size > 1 \", vamos[vamos['max_motif_size'] > 1]['mendelian_consistent_1'].sum()/vamos[vamos['max_motif_size'] > 1].shape[0])\n",
    "print(\"recall for vamos \", vamos.shape[0]/1243954)\n",
    "\n",
    "#### CLR \n",
    "# number of regions in vamos  1239659\n",
    "# Number of mendelian consistent regions for vamos  0.43918287206401113\n",
    "# Number of mendelian consistent off by one regions for vamos  0.7118239774002367\n",
    "# Number of mendelian consistent regions for vamos with motif size > 1  0.5369489043460071\n",
    "# Number of mendelian consistent regions off by one for vamos with motif size > 1  0.8087400432690204\n",
    "# recall for vamos  0.9965472999805459\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TandemTwister_MC =  tandemtwister['mendelian_consistent_0'].sum()/tandemtwister.shape[0]\n",
    "#TandemTwister_MC_motif =  tandemtwister['mendelian_consistent_0_motif'].sum()/tandemtwister.shape[0]\n",
    "Vamos_MC = vamos['mendelian_consistent_0'].sum()/vamos.shape[0]\n",
    "#Vamos_MC_motif = vamos['mendelian_consistent_0_motif'].sum()/vamos.shape[0]\n",
    "\n",
    "#TRGT_MC = TRGT['mendelian_consistent_0'].sum()/TRGT.shape[0]\n",
    "#TRGT_MC_motif = TRGT['mendelian_consistent_0_motif'].sum()/TRGT.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TandemTwister_MC_off_by_one = tandemtwister['mendelian_consistent_1'].sum()/tandemtwister.shape[0]\n",
    "#TandemTwister_MC_off_by_one_motif = tandemtwister['mendelian_consistent_1_motif'].sum()/tandemtwister.shape[0]\n",
    "Vamos_MC_off_by_one = vamos['mendelian_consistent_1'].sum()/vamos.shape[0]\n",
    "#Vamos_MC_off_by_one_motif = vamos['mendelian_consistent_1_motif'].sum()/vamos.shape[0]\n",
    "\n",
    "#TRGT_MC_off_by_one = TRGT['mendelian_consistent_1'].sum()/TRGT.shape[0]\n",
    "##TRGT_MC_off_by_one_motif = TRGT['mendelian_consistent_1_motif'].sum()/TRGT.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "recall_TandemTwister_MC = tandemtwister.shape[0]/1243954\n",
    "recall_Vamos_MC = vamos.shape[0]/1243954\n",
    "#recall_TRGT_MC = TRGT.shape[0]/1243954\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"TandemTwister_MC =\", TandemTwister_MC)\n",
    "print(\"Vamos_MC =\", Vamos_MC)\n",
    "#print(\"TRGT_MC \", TRGT_MC)\n",
    "\n",
    "\n",
    "print(\"TandemTwister_MC_off_by_one= \", TandemTwister_MC_off_by_one)\n",
    "print(\"Vamos_MC_off_by_one =\", Vamos_MC_off_by_one)\n",
    "#print(\"TRGT_MC_off_by_one \", TRGT_MC_off_by_one)\n",
    "\n",
    "print(\"recall_TandemTwister_MC = \", recall_TandemTwister_MC)\n",
    "print(\"recall_Vamos_MC =\", recall_Vamos_MC)\n",
    "#print(\"recall_TRGT_MC \", recall_TRGT_MC)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "TandemTwister_MC_CCS = 0.9487292577123414\n",
    "Vamos_MC_CCS = 0.812866225853674\n",
    "TRGT_MC_CCS = 0.9535301374278667\n",
    "\n",
    "TandemTwister_MC_off_by_one_CCS = 0.9847607682507942\n",
    "Vamos_MC_off_by_one_CCS = 0.9566472748871735\n",
    "TRGT_MC_off_by_one_CCS = 0.9890338847628842\n",
    "\n",
    "recall_TandemTwister_MC_CCS = 0.997\n",
    "recall_Vamos_MC_CCS = 0.993\n",
    "recall_TRGT_MC_CCS = 0.995\n",
    "\n",
    "TandemTwister_MC_CLR = 0.5178540904033155\n",
    "Vamos_MC_CLR = 0.43918287206401113\n",
    "\n",
    "TandemTwister_MC_off_by_one_CLR = 0.8182814339391399\n",
    "Vamos_MC_off_by_one_CLR = 0.7118239774002367\n",
    "\n",
    "recall_TandemTwister_MC_CLR = 0.9969982812869286\n",
    "recall_Vamos_MC_CLR = 0.9965472999805459\n",
    "\n",
    "TandemTwister_MC_ONT = 0.6353058328361098\n",
    "Vamos_MC_ONT = 0.49914080605957833\n",
    "TandemTwister_MC_off_by_one_ONT=  0.8663083968973246\n",
    "Vamos_MC_off_by_one_ONT = 0.7762518471862206\n",
    "recall_TandemTwister_MC_ONT =  0.9969982812869286\n",
    "recall_Vamos_MC_ONT = 0.9955110880305863\n",
    "\n",
    "\n",
    "\n",
    "# Round the values to 2 or 3 decimal places for better readability\n",
    "Vamos_MC_CCS = round(Vamos_MC_CCS, 3)\n",
    "TRGT_MC_CCS = round(TRGT_MC_CCS, 3)\n",
    "TandemTwister_MC_CCS = round(TandemTwister_MC_CCS, 3)\n",
    "\n",
    "Vamos_MC_off_by_one_CCS = round(Vamos_MC_off_by_one_CCS, 3)\n",
    "TRGT_MC_off_by_one_CCS = round(TRGT_MC_off_by_one_CCS, 3)\n",
    "TandemTwister_MC_off_by_one_CCS = round(TandemTwister_MC_off_by_one_CCS, 3)\n",
    "\n",
    "recall_Vamos_MC_CCS = round(recall_Vamos_MC_CCS, 3)\n",
    "recall_TRGT_MC_CCS = round(recall_TRGT_MC_CCS, 3)\n",
    "recall_TandemTwister_MC_CCS = round(recall_TandemTwister_MC_CCS, 3)\n",
    "\n",
    "Vamos_MC_CLR = round(Vamos_MC_CLR, 3)\n",
    "TandemTwister_MC_CLR = round(TandemTwister_MC_CLR, 3)\n",
    "\n",
    "Vamos_MC_off_by_one_CLR = round(Vamos_MC_off_by_one_CLR, 3)\n",
    "TandemTwister_MC_off_by_one_CLR = round(TandemTwister_MC_off_by_one_CLR, 3)\n",
    "\n",
    "recall_Vamos_MC_CLR = round(recall_Vamos_MC_CLR, 3)\n",
    "recall_TandemTwister_MC_CLR = round(recall_TandemTwister_MC_CLR, 3)\n",
    "\n",
    "\n",
    "Vamos_MC_ONT = round(Vamos_MC_ONT, 3)\n",
    "TandemTwister_MC_ONT = round(TandemTwister_MC_ONT, 3)\n",
    "\n",
    "Vamos_MC_off_by_one_ONT = round(Vamos_MC_off_by_one_ONT, 3)\n",
    "TandemTwister_MC_off_by_one_ONT = round(TandemTwister_MC_off_by_one_ONT, 3)\n",
    "\n",
    "recall_Vamos_MC_ONT = round(recall_Vamos_MC_ONT, 3)\n",
    "recall_TandemTwister_MC_ONT = round(recall_TandemTwister_MC_ONT, 3)\n",
    "# Data for CCS\n",
    "tools_CCS = ['Vamos', 'TRGT', 'TandemTwister']\n",
    "metrics_CCS = ['MC', 'MC Off by One', 'Recall']\n",
    "values_CCS = [\n",
    "    [Vamos_MC_CCS, TRGT_MC_CCS, TandemTwister_MC_CCS],  # Mendelian Consistency\n",
    "    [Vamos_MC_off_by_one_CCS, TRGT_MC_off_by_one_CCS, TandemTwister_MC_off_by_one_CCS],  # Off by One\n",
    "    [recall_Vamos_MC_CCS, recall_TRGT_MC_CCS, recall_TandemTwister_MC_CCS]  # Recall\n",
    "]\n",
    "\n",
    "# Data for CLR\n",
    "tools_CLR = ['Vamos', 'TandemTwister']\n",
    "metrics_CLR = ['MC', 'MC Off by One', 'Recall']\n",
    "values_CLR = [\n",
    "    [Vamos_MC_CLR, TandemTwister_MC_CLR],  # Mendelian Consistency\n",
    "    [Vamos_MC_off_by_one_CLR, TandemTwister_MC_off_by_one_CLR],  # Off by One\n",
    "    [recall_Vamos_MC_CLR, recall_TandemTwister_MC_CLR]  # Recall\n",
    "]\n",
    "\n",
    "# Data for ONT\n",
    "tools_ONT = ['Vamos', 'TandemTwister']\n",
    "metrics_ONT = ['MC', 'MC Off by One', 'Recall']\n",
    "values_ONT = [\n",
    "    [Vamos_MC_ONT, TandemTwister_MC_ONT],  # Mendelian Consistency\n",
    "    [Vamos_MC_off_by_one_ONT, TandemTwister_MC_off_by_one_ONT],  # Off by One\n",
    "    [recall_Vamos_MC_ONT, recall_TandemTwister_MC_ONT]  # Recall\n",
    "]\n",
    "\n",
    "\n",
    "# Create a subplot figure with 2 columns\n",
    "fig = make_subplots(rows=2, cols=2, subplot_titles=(\"CCS Data\", \"CLR Data\", \"ONT Data\"))\n",
    "colors = ['#1f77b4', '#ff7f0e', '#2ca02c']  \n",
    "# Add traces for CCS\n",
    "for i, metric in enumerate(metrics_CCS):\n",
    "    fig.add_trace(go.Bar(\n",
    "        name=metric,\n",
    "        x=tools_CCS,\n",
    "        y=values_CCS[i],\n",
    "        text=values_CCS[i],\n",
    "        textposition='inside',\n",
    "        marker_color=colors[i],\n",
    "        width=0.2,\n",
    "        showlegend=True \n",
    "    ), row=1, col=1)\n",
    "\n",
    "# Add traces for CLR\n",
    "for i, metric in enumerate(metrics_CLR):\n",
    "    fig.add_trace(go.Bar(\n",
    "        name=metric,\n",
    "        x=tools_CLR,\n",
    "        y=values_CLR[i],\n",
    "        text=values_CLR[i],\n",
    "        textposition='inside',\n",
    "        marker_color=colors[i],\n",
    "        width=0.2,\n",
    "        showlegend=False\n",
    "\n",
    "    ), row=1, col=2)\n",
    "\n",
    "# Add traces for ONT\n",
    "for i, metric in enumerate(metrics_ONT):\n",
    "    fig.add_trace(go.Bar(\n",
    "        name=metric,\n",
    "        x=tools_ONT,\n",
    "        y=values_ONT[i],\n",
    "        text=values_ONT[i],\n",
    "        textposition='inside',\n",
    "        marker_color=colors[i],\n",
    "        width=0.2,\n",
    "        showlegend=False\n",
    "    ), row=2, col=1)\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=\"Mendelian Consistency and Recall for Different Tools on Hifi Data (CCS vs CLR)\",\n",
    "    barmode='group',\n",
    "    xaxis_title=\"\",\n",
    "    yaxis_title=\"Scores\",\n",
    "    legend_title=\"\",\n",
    "    height=700,\n",
    "    width=1200,\n",
    "    font=dict(size=14),\n",
    "    title_font=dict(size=18, family='Arial'),\n",
    "    plot_bgcolor='rgba(240, 240, 240, 1)',\n",
    "    showlegend=True,  # Enable single legend for all subplots\n",
    "    legend=dict(\n",
    "        orientation='h',\n",
    "        yanchor=\"bottom\",\n",
    "        y=1.02,\n",
    "        xanchor=\"center\",\n",
    "        x=0.5\n",
    "    )\n",
    ")\n",
    "# remove the legend from the second subplot\n",
    "#fig.update_traces(showlegend=False, selector=dict(row=1, col=2))\n",
    "# Show the plot\n",
    "fig.show()\n",
    "\n",
    "# Save the plot as a PDF\n",
    "# fig.write_image(\"/confidential/tGenVar/Lion/TandemTwist/MC/paper_figures/mendelian_consistency_recall_CCS_CLR_ONT.pdf\", width=1200, height=500)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "\n",
    "# Combine tools and technologies for x-axis\n",
    "tools_technologies = [\n",
    "    'ONT_vamos', 'ONT_TandemTwister',  \n",
    "    'CLR_vamos', 'CLR_TandemTwister', \n",
    "    'CCS_vamos', 'CCS_TandemTwister', 'CCS_TRGT'\n",
    "]\n",
    "\n",
    "TandemTwister_MC_CCS = 0.9487292577123414\n",
    "Vamos_MC_CCS = 0.812866225853674\n",
    "TRGT_MC_CCS = 0.9535301374278667\n",
    "\n",
    "TandemTwister_MC_off_by_one_CCS = 0.9847607682507942\n",
    "Vamos_MC_off_by_one_CCS = 0.9566472748871735\n",
    "TRGT_MC_off_by_one_CCS = 0.9890338847628842\n",
    "\n",
    "recall_TandemTwister_MC_CCS = 0.997\n",
    "recall_Vamos_MC_CCS = 0.993\n",
    "recall_TRGT_MC_CCS = 0.995\n",
    "\n",
    "TandemTwister_MC_CLR = 0.5178540904033155\n",
    "Vamos_MC_CLR = 0.43918287206401113\n",
    "\n",
    "TandemTwister_MC_off_by_one_CLR = 0.8182814339391399\n",
    "Vamos_MC_off_by_one_CLR = 0.7118239774002367\n",
    "\n",
    "recall_TandemTwister_MC_CLR = 0.9969982812869286\n",
    "recall_Vamos_MC_CLR = 0.9965472999805459\n",
    "\n",
    "TandemTwister_MC_ONT = 0.6353058328361098\n",
    "Vamos_MC_ONT = 0.49914080605957833\n",
    "TandemTwister_MC_off_by_one_ONT=  0.8663083968973246\n",
    "Vamos_MC_off_by_one_ONT = 0.7762518471862206\n",
    "recall_TandemTwister_MC_ONT =  0.9969982812869286\n",
    "recall_Vamos_MC_ONT = 0.9955110880305863\n",
    "\n",
    "\n",
    "\n",
    "# Round the values to 2 or 3 decimal places for better readability\n",
    "Vamos_MC_CCS = round(Vamos_MC_CCS, 3)\n",
    "TRGT_MC_CCS = round(TRGT_MC_CCS, 3)\n",
    "TandemTwister_MC_CCS = round(TandemTwister_MC_CCS, 3)\n",
    "\n",
    "Vamos_MC_off_by_one_CCS = round(Vamos_MC_off_by_one_CCS, 3)\n",
    "TRGT_MC_off_by_one_CCS = round(TRGT_MC_off_by_one_CCS, 3)\n",
    "TandemTwister_MC_off_by_one_CCS = round(TandemTwister_MC_off_by_one_CCS, 3)\n",
    "\n",
    "recall_Vamos_MC_CCS = round(recall_Vamos_MC_CCS, 3)\n",
    "recall_TRGT_MC_CCS = round(recall_TRGT_MC_CCS, 3)\n",
    "recall_TandemTwister_MC_CCS = round(recall_TandemTwister_MC_CCS, 3)\n",
    "\n",
    "Vamos_MC_CLR = round(Vamos_MC_CLR, 3)\n",
    "TandemTwister_MC_CLR = round(TandemTwister_MC_CLR, 3)\n",
    "\n",
    "Vamos_MC_off_by_one_CLR = round(Vamos_MC_off_by_one_CLR, 3)\n",
    "TandemTwister_MC_off_by_one_CLR = round(TandemTwister_MC_off_by_one_CLR, 3)\n",
    "\n",
    "recall_Vamos_MC_CLR = round(recall_Vamos_MC_CLR, 3)\n",
    "recall_TandemTwister_MC_CLR = round(recall_TandemTwister_MC_CLR, 3)\n",
    "\n",
    "\n",
    "Vamos_MC_ONT = round(Vamos_MC_ONT, 3)\n",
    "TandemTwister_MC_ONT = round(TandemTwister_MC_ONT, 3)\n",
    "\n",
    "Vamos_MC_off_by_one_ONT = round(Vamos_MC_off_by_one_ONT, 3)\n",
    "TandemTwister_MC_off_by_one_ONT = round(TandemTwister_MC_off_by_one_ONT, 3)\n",
    "\n",
    "recall_Vamos_MC_ONT = round(recall_Vamos_MC_ONT, 3)\n",
    "recall_TandemTwister_MC_ONT = round(recall_TandemTwister_MC_ONT, 3)\n",
    "\n",
    "MC_values = [Vamos_MC_CLR, TandemTwister_MC_CLR, Vamos_MC_ONT, TandemTwister_MC_ONT, Vamos_MC_CCS, TandemTwister_MC_CCS, TRGT_MC_CCS]\n",
    "MC_off_by_one_values = [Vamos_MC_off_by_one_CLR, TandemTwister_MC_off_by_one_CLR, Vamos_MC_off_by_one_ONT, TandemTwister_MC_off_by_one_ONT, Vamos_MC_off_by_one_CCS, TandemTwister_MC_off_by_one_CCS, TRGT_MC_off_by_one_CCS]\n",
    "recall_values = [recall_Vamos_MC_CLR, recall_TandemTwister_MC_CLR, recall_Vamos_MC_ONT, recall_TandemTwister_MC_ONT, recall_Vamos_MC_CCS, recall_TandemTwister_MC_CCS, recall_TRGT_MC_CCS]\n",
    "\n",
    "# Colors for the bars\n",
    "colors = ['#1f77b4', '#ff7f0e', '#2ca02c']  # Blue, orange, and green\n",
    "\n",
    "# Create the figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add Mendelian consistency bars\n",
    "fig.add_trace(go.Bar(\n",
    "    name=\"MC\",\n",
    "    x=tools_technologies,\n",
    "    y=MC_values,\n",
    "    text=MC_values,\n",
    "    textposition='inside',\n",
    "    marker_color=colors[0],\n",
    "    width=0.2\n",
    "))\n",
    "\n",
    "# Add MC off by one bars\n",
    "fig.add_trace(go.Bar(\n",
    "    name=\"MC_off_by_1\",\n",
    "    x=tools_technologies,\n",
    "    y=MC_off_by_one_values,\n",
    "    text=MC_off_by_one_values,\n",
    "    textposition='inside',\n",
    "    marker_color=colors[1],\n",
    "    width=0.2\n",
    "))\n",
    "\n",
    "# Add Recall bars\n",
    "fig.add_trace(go.Bar(\n",
    "    name=\"recall\",\n",
    "    x=tools_technologies,\n",
    "    y=recall_values,\n",
    "    text=recall_values,\n",
    "    textposition='inside',\n",
    "    marker_color=colors[2],\n",
    "    width=0.2\n",
    "))\n",
    "\n",
    "# Update layout\n",
    "fig.update_layout(\n",
    "    title=\"Mendelian Consistency and Recall for Different Tools across Different Technologies\",\n",
    "    barmode='group',\n",
    "    yaxis_title=\"Mendelian Consistency and Recall (Percentage)\",\n",
    "    yaxis_range=[0, 1],  # Set range for percentages\n",
    "    height=600,\n",
    "    width=1000,\n",
    "    font=dict(size=14),\n",
    "    title_font=dict(size=18, family='Arial'),\n",
    "    plot_bgcolor='rgba(240, 240, 240, 1)',\n",
    "    showlegend=True,\n",
    "    legend=dict(\n",
    "        orientation='h',\n",
    "        yanchor=\"bottom\",\n",
    "        y=1.02,\n",
    "        xanchor=\"center\",\n",
    "        x=0.5\n",
    "    )\n",
    ")\n",
    "\n",
    "# Rotate x-axis labels\n",
    "fig.update_xaxes(tickangle=-45)\n",
    "\n",
    "# Show the plot\n",
    "fig.show()\n",
    "\n",
    "# Save the plot as a PDF\n",
    "fig.write_image(\"/confidential/tGenVar/Lion/TandemTwist/MC/paper_figures/mendelian_consistency_recall_by_technology.pdf\", width=1000, height=600)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "\n",
    "# Define the data for the comparison table\n",
    "data = {\n",
    "    \"Features\": [\n",
    "        \"Supports CCS_input\", \n",
    "        \"Supports ONT_input\", \n",
    "        \"Supports CLR_input\", \n",
    "        \"Supports assembly input\", \n",
    "        \"Run in parallel\", \n",
    "    ],\n",
    "    \"TandemTwister\": [\"✅\", \"✅\", \"✅\", \"✅\", \"✅\"],\n",
    "    \"TRGT\": [\"❌\", \"❌\", \"❌\", \"✅\", \"✅\",],\n",
    "    \"Vamos\": [\"✅\", \"✅\", \"✅\", \"✅\", \"❌\"]  # Adding Vamos for comparison\n",
    "}\n",
    "\n",
    "# Create a Plotly table\n",
    "fig = go.Figure(data=[go.Table(\n",
    "    header=dict(values=[\"Features\", \"TandemTwister\", \"Tandem-genotypes\", \"TRGT\"],\n",
    "                fill_color='paleturquoise',\n",
    "                align='left'),\n",
    "    cells=dict(values=[data[\"Features\"], data[\"TandemTwister\"],  data[\"TRGT\"], data[\"Vamos\"]],\n",
    "               fill_color='lavender',\n",
    "               align='left'))\n",
    "])\n",
    "\n",
    "# Update layout for better visualization\n",
    "fig.update_layout(\n",
    "    title=\"Comparison of Tandem Repeat Genotyping Tools\",\n",
    "    width=800,\n",
    "    height=400\n",
    ")\n",
    "\n",
    "# Show the plot\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tandemtwister_MC_per_motif_size =[tandemtwister[tandemtwister['max_motif_size'] == i]['mendelian_consistent_0'].sum()/tandemtwister[tandemtwister['max_motif_size'] == i].shape[0] for i in range(1, 11)]\n",
    "tandemtwister_MC_per_motif_size_off_by_one =[tandemtwister[tandemtwister['max_motif_size'] == i]['mendelian_consistent_1'].sum()/tandemtwister[tandemtwister['max_motif_size'] == i].shape[0] for i in range(1, 11)]\n",
    "\n",
    "vamos_MC_per_motif_size =[vamos[vamos['max_motif_size'] == i]['mendelian_consistent_0'].sum()/vamos[vamos['max_motif_size'] == i].shape[0] for i in range(1, 11)]\n",
    "vamos_MC_per_motif_size_off_by_one =[vamos[vamos['max_motif_size'] == i]['mendelian_consistent_1'].sum()/vamos[vamos['max_motif_size'] == i].shape[0] for i in range(1, 11)]\n",
    "\n",
    "print(\"tandemtwister_MC_per_motif_size = \", tandemtwister_MC_per_motif_size)\n",
    "print(\"tandemtwister_MC_per_motif_size_off_by_one = \", tandemtwister_MC_per_motif_size_off_by_one)\n",
    "print(\"vamos_MC_per_motif_size = \", vamos_MC_per_motif_size)\n",
    "print(\"vamos_MC_per_motif_size_off_by_one = \", vamos_MC_per_motif_size_off_by_one)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "# Data (as placeholders, since actual data isn't provided)\n",
    "#### CCS results\n",
    "# tandemtwister_MC_per_motif_size=  [0.7689604669209666, 0.9506249581025162, 0.9783060261038601, 0.989989024772656, 0.9772590821474841, 0.975196104301101, 0.987943176135058, 0.9869544633620312, 0.982550995330548, 0.9795629820051414]\n",
    "# trgt_MC_per_motif_size = [0.8325180424243753, 0.9351182774699118, 0.9786415111130887, 0.984628670120898, 0.9802907847812784, 0.9742530006127378, 0.9821694511386099, 0.972425663452588, 0.9614167135621835, 0.9582984748053285]\n",
    "# vamos_MC_per_motif_size = [0.47793868197510825, 0.7212122468439709, 0.822845786335698, 0.912506900086744, 0.910878375194383, 0.9202378362700312, 0.9611614900086881, 0.96648192135128, 0.9600578156948459, 0.9545043298436087]\n",
    "# tandemtwister_MC_per_motif_size_off_by_one = [0.945465773337851, 0.9901518060100798, 0.9934351569008609, 0.995931326434619, 0.9866440556567901, 0.9846834744644614, 0.99266047605531, 0.9922673633637846, 0.9900291401888845, 0.9873393316195372]\n",
    "# trgt_MC_per_motif_size_off_by_one = [0.9887248557702853, 0.9899787613211923, 0.9949941041671302, 0.9954152928246193, 0.9901875962776172, 0.9833599653983396, 0.9871517153610788, 0.9780107838488153, 0.9683806978052898, 0.9660853336765558]\n",
    "# vamos_MC_per_motif_size_off_by_one = [0.8717629763654868, 0.9447461702701022, 0.9628713197067611, 0.9855985332387036, 0.981680882145045, 0.9840959079592971, 0.9939590573414422, 0.9951966217999472, 0.9917154339702461, 0.9917926845030374]\n",
    "\n",
    "## CLR results\n",
    "# tandemtwister_MC_per_motif_size =  [0.12560523755915753, 0.2298907313624757, 0.2530630380449875, 0.27929601756036376, 0.5374714290228463, 0.5476047688727901, 0.7719340929673033, 0.7358454174045694, 0.6719446687497805, 0.6810411311053984]\n",
    "# tandemtwister_MC_per_motif_size_off_by_one =  [0.542320177499918, 0.6687204051410498, 0.720410996945293, 0.7955785512699906, 0.8920464719451437, 0.8931801280974885, 0.9556925239582066, 0.9538496607108414, 0.9358213671312713, 0.9282133676092545]\n",
    "# vamos_MC_per_motif_size =  [0.1366035741419747, 0.10479382616431837, 0.19260960882121292, 0.36124891723757774, 0.37433570051262754, 0.4090596454816406, 0.5239567233384853, 0.5430016863406408, 0.5409207610874688, 0.5432345360824742]\n",
    "# vamos_MC_per_motif_size_off_by_one =  [0.41187526839097544, 0.41807874537065776, 0.5430427933840903, 0.7756220962280495, 0.779523115270658, 0.7969061093480385, 0.8775861601453402, 0.883080382237212, 0.868568212991946, 0.8728737113402062]\n",
    "\n",
    "## ONT results\n",
    "tandemtwister_MC_per_motif_size =  [0.32999245843944347, 0.2699233961643235, 0.5114801444043321, 0.6720131702728128, 0.6850398677045261, 0.685525463573776, 0.8191206087885055, 0.7845733022391331, 0.744268511041674, 0.7339974293059126]\n",
    "tandemtwister_MC_per_motif_size_off_by_one =  [0.6744778289049436, 0.6567453425233559, 0.8327242432657596, 0.9259799310128567, 0.9277746763711435, 0.9281910427711277, 0.9718719165213631, 0.9701039785379881, 0.9574483024962258, 0.9480719794344473]\n",
    "vamos_MC_per_motif_size =  [0.1844932290129251, 0.1526510684954178, 0.2995403204623063, 0.4920681782396473, 0.4839695733449334, 0.5065281397818322, 0.6072721356814577, 0.6194231715028455, 0.6105777684003235, 0.6121278515272587]\n",
    "vamos_MC_per_motif_size_off_by_one =  [0.5016500446092462, 0.5057613934917975, 0.6918308379301287, 0.8697055581798142, 0.8538509470120037, 0.8635968722849696, 0.9247027561990754, 0.9290908452188575, 0.9226360023912509, 0.9177729088800103]\n",
    "\n",
    "# Create Plotly figure\n",
    "fig = go.Figure()\n",
    "\n",
    "# Add traces for each dataset\n",
    "fig.add_trace(go.Scatter(x=list(range(1, 11)), y=tandemtwister_MC_per_motif_size, mode='lines+markers', name='TandemTwister', marker=dict(symbol='circle')))\n",
    "fig.add_trace(go.Scatter(x=list(range(1, 11)), y=tandemtwister_MC_per_motif_size_off_by_one, mode='lines+markers', name='TandemTwister off by one', marker=dict(symbol='x')))\n",
    "# fig.add_trace(go.Scatter(x=list(range(1, 11)), y=trgt_MC_per_motif_size, mode='lines+markers', name='TRGT', marker=dict(symbol='circle')))\n",
    "# fig.add_trace(go.Scatter(x=list(range(1, 11)), y=trgt_MC_per_motif_size_off_by_one, mode='lines+markers', name='TRGT off by one', marker=dict(symbol='x')))\n",
    "fig.add_trace(go.Scatter(x=list(range(1, 11)), y=vamos_MC_per_motif_size, mode='lines+markers', name='Vamos', marker=dict(symbol='circle')))\n",
    "fig.add_trace(go.Scatter(x=list(range(1, 11)), y=vamos_MC_per_motif_size_off_by_one, mode='lines+markers', name='Vamos off by one', marker=dict(symbol='x')))\n",
    "\n",
    "\n",
    "# Update the layout without the invalid grid property\n",
    "fig.update_layout(\n",
    "    title='Mendelian Consistency per Motif Size',\n",
    "    xaxis_title='Motif Size',\n",
    "    yaxis_title='Mendelian Consistency',\n",
    "    legend=dict(\n",
    "        orientation=\"h\",\n",
    "        yanchor=\"bottom\",\n",
    "        y=1.0,\n",
    "        xanchor=\"center\",\n",
    "        x=0.5\n",
    "    ),\n",
    "    margin=dict(t=80)\n",
    ")\n",
    "# make the y axis from 0 to 1\n",
    "fig.update_yaxes(range=[0, 1.1])\n",
    "\n",
    "# Show the plot\n",
    "fig.show()\n",
    "# save the plot as a pdf\n",
    "fig.write_image(\"/confidential/tGenVar/Lion/TandemTwist/MC/paper_figures/mendelian_consistency_per_motif_size_ONT.pdf\", width=900, height=500)\n",
    "### CCS results\n",
    "# tandemtwister_MC_per_motif_size =  [0.7689604669209666, 0.9506249581025162, 0.9783060261038601, 0.989989024772656, 0.9772590821474841, 0.975196104301101, 0.987943176135058, 0.9869544633620312, 0.982550995330548, 0.9795629820051414]\n",
    "# trgt_MC_per_motif_size = [0.8325180424243753, 0.9351182774699118, 0.9786415111130887, 0.984628670120898, 0.9802907847812784, 0.9742530006127378, 0.9821694511386099, 0.972425663452588, 0.9614167135621835, 0.9582984748053285]\n",
    "# vamos_MC_per_motif_size = [0.47793868197510825, 0.7212122468439709, 0.822845786335698, 0.912506900086744, 0.910878375194383, 0.9202378362700312, 0.9611614900086881, 0.96648192135128, 0.9600578156948459, 0.9545043298436087]\n",
    "# tandemtwister_MC_per_motif_size_off_by_one = [0.945465773337851, 0.9901518060100798, 0.9934351569008609, 0.995931326434619, 0.9866440556567901, 0.9846834744644614, 0.99266047605531, 0.9922673633637846, 0.9900291401888845, 0.9873393316195372]\n",
    "# trgt_MC_per_motif_size_off_by_one = [0.9887248557702853, 0.9899787613211923, 0.9949941041671302, 0.9954152928246193, 0.9901875962776172, 0.9833599653983396, 0.9871517153610788, 0.9780107838488153, 0.9683806978052898, 0.9660853336765558]\n",
    "# vamos_MC_per_motif_size_off_by_one = [0.8717629763654868, 0.9447461702701022, 0.9628713197067611, 0.9855985332387036, 0.981680882145045, 0.9840959079592971, 0.9939590573414422, 0.9951966217999472, 0.9917154339702461, 0.9917926845030374]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib_venn import venn3\n",
    "\n",
    "def plot_MC_by_motif_size(TRGT, tandemtwister, vamos):\n",
    "    rates = [\n",
    "        TRGT['mendelian_consistent'].sum() / TRGT.shape[0],\n",
    "        tandemtwister['mendelian_consistent'].sum() / tandemtwister.shape[0],\n",
    "        vamos['mendelian_consistent'].sum() / vamos.shape[0]\n",
    "    ]\n",
    "    tools = ['TRGT', 'TandemTwister', 'VAMOS']\n",
    "    \n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.barplot(x=tools, y=                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           _               \n",
    "    return consistency_rate\n",
    "\n",
    "\n",
    "def plot_mendelian_consistency_rate(tandemtwister, TRGT, vamos):\n",
    "    rates = [\n",
    "        mendelian_consistency_rate(tandemtwister, 'TandemTwister'),\n",
    "        mendelian_consistency_rate(TRGT, 'TRGT'),\n",
    "        mendelian_consistency_rate(vamos, 'VAMOS')\n",
    "    ]\n",
    "    tools = ['TandemTwister', 'TRGT', 'VAMOS']\n",
    "    \n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.barplot(x=tools, y=rates)\n",
    "    plt.ylabel('Mendelian Consistency Rate (%)')\n",
    "    plt.title('Mendelian Consistency Rate by Tool')\n",
    "    plt.show()\n",
    "\n",
    "def plot_venn_diagram(tandemtwister, TRGT, vamos):\n",
    "\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    regions_tandemtwister = set(tandemtwister[tandemtwister['mendelian_consistent']]['region'])\n",
    "    regions_TRGT = set(TRGT[TRGT['mendelian_consistent']]['region'])\n",
    "    regions_vamos = set(vamos[vamos['mendelian_consistent']]['region'])\n",
    "    plt.figure(figsize=(8, 8))\n",
    "    venn3(\n",
    "        [regions_tandemtwister, regions_TRGT, regions_vamos],\n",
    "        ('TandemTwister', 'TRGT', 'VAMOS')\n",
    "    )\n",
    "\n",
    "    plt.title('Number of Regions Called by Each Tool')\n",
    "    plt.show()\n",
    "def calcluate_error_rate_by_region(row):\n",
    "    # calculate the error rate by assiging the hap1_child to either hap1_father or hap2_father and hap2_child to either hap1_mother or hap2_mother based on the minimum error rate\n",
    "    if (row[\"chrom\"] != \"X\" and row[\"chrom\"] != \"Y\"):\n",
    "        error_rate_h1_father_h1_mother = abs(row['CN_H1_child'] - row['CN_H1_father']) + abs(row['CN_H1_child'] - row['CN_H1_mother'])\n",
    "        error_rate_h1_father_h2_mother = abs(row['CN_H1_child'] - row['CN_H1_father']) + abs(row['CN_H1_child'] - row['CN_H2_mother'])\n",
    "        error_rate_h2_father_h1_mother = abs(row['CN_H2_child'] - row['CN_H2_father']) + abs(row['CN_H2_child'] - row['CN_H1_mother'])\n",
    "        error_rate_h2_father_h2_mother = abs(row['CN_H2_child'] - row['CN_H2_father']) + abs(row['CN_H2_child'] - row['CN_H2_mother'])\n",
    "\n",
    "        error_rate = min(\n",
    "            error_rate_h1_father_h1_mother,\n",
    "            error_rate_h1_father_h2_mother,\n",
    "            error_rate_h2_father_h1_mother,\n",
    "            error_rate_h2_father_h2_mother\n",
    "        )\n",
    "\n",
    "        return error_rate\n",
    "    else:\n",
    "        if row[\"chrom\"] == \"X\":\n",
    "            error_rate_h1_father_h1_mother = abs(row['CN_H1_child'] - row['CN_H1_father']) + abs(row['CN_H1_child'] - row['CN_H1_mother'])\n",
    "            error_rate_h1_father_h2_mother = abs(row['CN_H1_child'] - row['CN_H1_father']) + abs(row['CN_H1_child'] - row['CN_H2_mother'])\n",
    "            error_rate = min(\n",
    "                error_rate_h1_father_h1_mother,\n",
    "                error_rate_h1_father_h2_mother\n",
    "            )\n",
    "            return error_rate\n",
    "        else:\n",
    "            error_rate_h1_father_h1_mother = abs(row['CN_H1_child'] - row['CN_H1_father']) + abs(row['CN_H1_child'] - row['CN_H1_mother'])\n",
    "            error_rate = error_rate_h1_father_h1_mother\n",
    "            return error_rate\n",
    "        \n",
    "def calculate_error_rate(df):\n",
    "    df['error_rate'] = df.apply(calcluate_error_rate_by_region, axis=1)\n",
    "    return df\n",
    "\n",
    "def plot_error_rate(tandemtwister, TRGT, vamos):\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.histplot(tandemtwister['error_rate'], bins=100, color='blue', label='TandemTwister', kde=True)\n",
    "    sns.histplot(TRGT['error_rate'], bins=100, color='red', label='TRGT', kde=True)\n",
    "    #sns.histplot(vamos['error_rate'], bins=100, color='green', label='VAMOS', kde=True)\n",
    "    plt.xlabel('Error Rate')\n",
    "    plt.ylabel('Number of Regions')\n",
    "    plt.title('Error Rate Distribution by Tool')\n",
    "    plt.legend()\n",
    "    plt.show()\n",
    "\n",
    "tandemtwister = calculate_error_rate(tandemtwister)\n",
    "TRGT = calculate_error_rate(TRGT)\n",
    "# vamos = calculate_error_rate(vamos)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "plot_mendelian_consistency_rate(tandemtwister, TRGT, vamos)\n",
    "plot_venn_diagram(tandemtwister, TRGT, vamos)\n",
    "plot_error_rate(tandemtwister, TRGT, vamos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TRGT "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tandemtwister_nan = tandemtwister[tandemtwister.isna().any(axis=1)]\n",
    "tandemtwister_nan.loc[tandemtwister[\"error_rate\"] == tandemtwister[\"error_rate\"].max()]\n",
    "#N_H1=4821;CN_H2=4839"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import FancyBboxPatch\n",
    "from matplotlib.font_manager import FontProperties\n",
    "\n",
    "prop = FontProperties(fname='/usr/share/fonts/truetype/Roboto-Italic.ttf', size=12)\n",
    "data = [\n",
    "    (1, 3, 'CTT'),\n",
    "    (5, 7, 'CTC'),\n",
    "    (8, 10, 'CTG'),\n",
    "    (11, 13, 'CTC'),\n",
    "    (14, 16, 'CTT'),\n",
    "    (17, 19, 'CTG'),\n",
    "    (20, 22, 'CTG'),\n",
    "    (23, 25, 'CTG'),\n",
    "    (26, 28, 'CTT'),\n",
    "    (29, 31, 'CTC'),\n",
    "    (32, 34, 'CAG'),\n",
    "    (35, 37, 'CTT'),\n",
    "    (41, 43, 'CTC'),\n",
    "    (44, 46, 'CTT'),\n",
    "    (47, 50, 'CATG'),\n",
    "    (51, 53, 'CTG'),\n",
    "]\n",
    "colors = {\n",
    "    'CTT': 'blue',\n",
    "    'CTC': 'green',\n",
    "    'CTG': 'orange',\n",
    "    'CAG': 'red',\n",
    "    'CATG': 'purple'\n",
    "}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(30, 10))  # Increase height of the figure\n",
    "\n",
    "# Adjust width and height scaling factors\n",
    "width_scale = 1  # One box per base pair\n",
    "height_scale = 0.1  # Height of each box\n",
    "\n",
    "# Track covered positions\n",
    "covered_positions = set()\n",
    "\n",
    "# Create base pair positions with colors based on motifs\n",
    "for (start, end, motif) in data:\n",
    "    for position in range(start, end + 1):\n",
    "        rect = FancyBboxPatch((position * width_scale, 0.5), width_scale, 0.5 * height_scale, \n",
    "                              boxstyle=\"round,pad=0.1\", \n",
    "                              edgecolor='black', \n",
    "                              facecolor=colors[motif],\n",
    "                              mutation_aspect=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        covered_positions.add(position)\n",
    "        ax.text(position * width_scale + 0.5, 0.50, motif[position - start], ha='center', va='center', color=\"black\", fontsize=15, fontproperties=prop)\n",
    "\n",
    "# Add gray boxes for base pairs not covered by any motifs\n",
    "for i in range(1, 55):\n",
    "    if i not in covered_positions:\n",
    "        rect = FancyBboxPatch((i * width_scale, 0.5), width_scale, 0.5 * height_scale, \n",
    "                              boxstyle=\"round,pad=0.1\", \n",
    "                              edgecolor='black', \n",
    "                              facecolor='gray',\n",
    "                              mutation_aspect=0.5)\n",
    "        ax.add_patch(rect)\n",
    "\n",
    "# Set the limits of the plot\n",
    "ax.set_xlim(0, 54 * width_scale)\n",
    "ax.set_ylim(0, 2)\n",
    "\n",
    "# Remove axes and ticks\n",
    "ax.axis('off')\n",
    "\n",
    "# Add an annotation on the plot that shows the total size of the region \n",
    "ax.annotate('Total size: 54 bp', xy=(0, 1.5), xytext=(0, 1.5), fontsize=12, fontproperties=prop)\n",
    "\n",
    "# Add an annotation that shows the position of each base pair underneath the boxes\n",
    "for i in range(0, 54):\n",
    "    ax.annotate(str(i+1), xy=(i * width_scale, 0.45), xytext=(i * width_scale, 0.35), fontsize=12, fontproperties=prop)\n",
    "\n",
    "# Add a more stylish legend\n",
    "handles = [FancyBboxPatch((0, 0), 1, 1, boxstyle=\"round,pad=0.1\", edgecolor='black', facecolor=colors[motif]) for motif in colors]\n",
    "labels = list(colors.keys())\n",
    "ax.legend(handles, labels, bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0., prop=prop)\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n",
    "#CTT,CTC,CTG,CAG,CATG;SVTYPE=VNTR;ALTANNO_H1=0,1,2,1,0,2,0,0,0,1,3,0,2,1,0,4,2 Vamos\n",
    "# CATG,CTT,CTC,CTG,CAG    1,2,3,2,1,3,3,3,1,2,4,1,2,1,0,3 TandemTwister\n",
    "#                         1,2,3,2,1,3,1,1,1,2,4,1,2,1,0,3\n",
    "                        # 1,2,3,2,1,3,1,1,1,2,4,1,0,0,3\n",
    "                        # 1,2,3,2,1,3,1,1,1,2,4,1,0,0,3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comparing runtimes on the assembly samples of hgsvc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import logomaker\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.animation import FuncAnimation\n",
    "from pptx import Presentation\n",
    "from pptx.util import Inches\n",
    "\n",
    "# Define your list of input motifs (sequences) - all approximately matching\n",
    "sequences = [\n",
    "    \"ATCGATCGAA\",\n",
    "    \"ATCGATCGAC\",\n",
    "    \"ATCGATCGAG\",\n",
    "    \"ATCGATCGTA\",\n",
    "    \"ATCGATCGTT\",\n",
    "    \"ATCGATCGGC\",\n",
    "    \"ATCGATCGCA\",\n",
    "    \"ATCGATCGTA\",\n",
    "]\n",
    "\n",
    "# Function to count nucleotides at each position\n",
    "def compute_position_counts(sequences):\n",
    "    counts_df = pd.DataFrame(0, index=['A', 'T', 'C', 'G'], columns=range(len(sequences[0])))\n",
    "    for seq in sequences:\n",
    "        for i, base in enumerate(seq):\n",
    "            counts_df.at[base, i] += 1\n",
    "    return counts_df\n",
    "\n",
    "# Initialize plot\n",
    "fig, ax = plt.subplots(figsize=(6, 4))\n",
    "ax.set_title(\"Motif Logo Evolution\")\n",
    "\n",
    "# Placeholder for sequence text\n",
    "sequence_text = ax.text(0.5, 1.1, \"\", ha=\"center\", va=\"center\", fontsize=12, transform=ax.transAxes)\n",
    "\n",
    "# Prepare animation function\n",
    "def update(frame):\n",
    "    ax.clear()\n",
    "    ax.set_title(\"Motif Logo Evolution\")\n",
    "    \n",
    "    # Display current sequence being added\n",
    "    current_sequence = sequences[frame]\n",
    "    sequence_text.set_text(f\"Added Sequence: {current_sequence}\")\n",
    "    \n",
    "    # Subset of sequences up to the current frame\n",
    "    subset_sequences = sequences[:frame + 1]\n",
    "    \n",
    "    # Calculate base frequencies\n",
    "    counts_df = compute_position_counts(subset_sequences)\n",
    "    \n",
    "    # Check if counts_df is valid\n",
    "    if counts_df.sum().sum() > 0:\n",
    "        logo = logomaker.Logo(counts_df, ax=ax)\n",
    "        logo.style_spines(visible=False)\n",
    "        logo.style_spines(spines=['left', 'bottom'], visible=True)\n",
    "        logo.style_xticks(rotation=0, fmt='%d', anchor=0)\n",
    "        logo.style_yaxis(gridline_style=':')\n",
    "\n",
    "# Create animation\n",
    "anim = FuncAnimation(fig, update, frames=len(sequences), interval=1000, repeat=False)\n",
    "\n",
    "# Save animation as MP4\n",
    "anim.save(\"motif_logo_evolution_with_sequences.mp4\", writer='ffmpeg', fps=1)\n",
    "\n",
    "# Create a PowerPoint presentation\n",
    "prs = Presentation()\n",
    "slide_layout = prs.slide_layouts[5]  # Use a blank slide layout\n",
    "\n",
    "# Add frames to the presentation\n",
    "for frame in range(len(sequences)):\n",
    "    # Create a new slide\n",
    "    slide = prs.slides.add_slide(slide_layout)\n",
    "    \n",
    "    # Add title\n",
    "    title = slide.shapes.title\n",
    "    title.text = \"Motif Logo Evolution\"\n",
    "    \n",
    "    # Add the animation frame as an image\n",
    "    anim.save(\"temp_frame.png\", writer='pillow', fps=1, dpi=100, savefig_kwargs={'bbox_inches': 'tight', 'pad_inches': 0.5})\n",
    "    slide.shapes.add_picture(\"temp_frame.png\", Inches(1), Inches(1), height=Inches(4.5))\n",
    "    \n",
    "    # Add the current sequence text\n",
    "    current_sequence = sequences[frame]\n",
    "    text_box = slide.shapes.add_textbox(Inches(1), Inches(5.7), Inches(8), Inches(0.5))\n",
    "    text_frame = text_box.text_frame\n",
    "    p = text_frame.add_paragraph()\n",
    "    p.text = f\"Added Sequence: {current_sequence}\"\n",
    "\n",
    "# Save the presentation\n",
    "prs.save(\"motif_logo_evolution_with_sequences.pptx\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
